<!DOCTYPE html>
<!--
Copyright (c) 2016 Matthew Vincent, The Churchill Lab

This software was developed by Gary Churchill's Lab.
(see http://research.jax.org/faculty/churchill)

This is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This software is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this software. If not, see <http://www.gnu.org/licenses/>.
//-->
<html>
<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title>Phenotype Viewer</title>
    <!-- Tell the browser to be responsive to screen width -->
    <meta content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no" name="viewport">
    <!-- Bootstrap 3.3.6 -->
    <link rel="stylesheet" href="{{g.URL_BASE}}/static/AdminLTE-2.3.7/bootstrap/css/bootstrap.min.css">
    <!-- Font Awesome -->
    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/font-awesome/4.5.0/css/font-awesome.min.css">
    <!-- Ionicons -->
    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/ionicons/2.0.1/css/ionicons.min.css">
    <!-- DataTables -->
    <link rel="stylesheet" href="{{g.URL_BASE}}/static/AdminLTE-2.3.7/plugins/datatables/dataTables.bootstrap.css">
    <!-- Theme style -->
    <link rel="stylesheet" href="{{g.URL_BASE}}/static/AdminLTE-2.3.7/dist/css/AdminLTE.css">
    <!-- AdminLTE Skins. We have chosen the skin-blue for this starter
          page. However, you can choose any other skin. Make sure you
          apply the skin class to the body tag so the changes take effect.
    -->
    <link rel="stylesheet" href="{{g.URL_BASE}}/static/AdminLTE-2.3.7/dist/css/skins/skin-blue-light.css">
    <link rel="stylesheet" href="{{g.URL_BASE}}/static/bootstrap-select/css/bootstrap-select.min.css">
    <link rel="stylesheet" href="{{g.URL_BASE}}/static/bootstrap-multiselect/css/bootstrap-multiselect.css">
    <link rel="stylesheet" href="{{g.URL_BASE}}/static/AdminLTE-2.3.7/plugins/ionslider/ion.rangeSlider.css">
    <!-- ion slider Nice -->
    <link rel="stylesheet" href="{{g.URL_BASE}}/static/AdminLTE-2.3.7/plugins/ionslider/ion.rangeSlider.skinNice.css">
    <link rel="stylesheet" href="//code.highcharts.com/css/highcharts.css">
    <link rel="stylesheet" href="{{g.URL_BASE}}/static/jquery-typeahead-2.9.0/dist/jquery.typeahead.min.css">

    <!-- local css -->
    <link rel="stylesheet" href="{{g.URL_BASE}}/static/css/style.css">

    <!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
  <script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
  <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
  <![endif]-->
</head>
<!--
BODY TAG OPTIONS:
=================
Apply one or more of the following classes to get the
desired effect
|---------------------------------------------------------|
| SKINS         | skin-blue                               |
|               | skin-black                              |
|               | skin-purple                             |
|               | skin-yellow                             |
|               | skin-red                                |
|               | skin-green                              |
|---------------------------------------------------------|
|LAYOUT OPTIONS | fixed                                   |
|               | layout-boxed                            |
|               | layout-top-nav                          |
|               | sidebar-collapse                        |
|               | sidebar-mini                            |
|---------------------------------------------------------|
-->
<body class="hold-transition skin-blue-light layout-top-nav">
<div class="wrapper">
    <header class="main-header">
        <nav class="navbar navbar-static-top">
            <div class="container">
                <div class="navbar-header">
                    <a href="#" class="navbar-brand">{{ CONF.MAIN_TITLE|safe }}</a>
                    <!--
                    <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar-collapse">
                        <i class="fa fa-bars"></i>
                    </button>
                    //-->
                </div>
            </div>
        </nav>
    </header>


    <!-- Content Wrapper. Contains page content -->
    <div class="content-wrapper">

        <!-- Content Header (Page header)
        <section class="content-header">
            <h1>
                QTL Viewer
                <small>Optional description</small>
            </h1>
            <ol class="breadcrumb">
                <li><a href="#"><i class="fa fa-dashboard"></i> Level</a></li>
                <li class="active">Here</li>
            </ol>
        </section>
        //-->

        <!-- Main content -->
        <section class="content">

            <div class="row">
                <div class="col-md-4">
                    <div class="row">
                        <div class="col-md-12">
                            <div id="search_box_div" class="box box-default">
                                <div class="box-header with-border">
                                    <h3 class="box-title"><i class="fa fa-search"></i> Search</h3>
                                    <div class="box-tools pull-right">
                                        <div id="div_btn_levels" class="btn-group" data-toggle="buttons">
                                            <!-- will be filled in via code //-->
                                        </div>
                                    </div>
                                </div>
                                <div class="box-body">
                                    <div class="nav-tabs-custom">
                                        <ul class="nav nav-tabs">
                                            <li class="active"><a href="#tab_1-1" data-toggle="tab">Search</a></li>
                                            <li><a href="#tab_2-2" data-toggle="tab">History</a></li>
                                        </ul>
                                    </div>
                                    <div class="tab-content">
                                        <div class="tab-pane active" id="tab_1-1">

                                            <div id="search_pheno" class="hidden">
                                                <form id="form-pheno" name="form-pheno">
                                                    <div class="typeahead__container">
                                                        <div class="typeahead__field">

                                                            <span class="typeahead__query">
                                                                <input class="js-typeahead-pheno" name="pheno[query]" type="search" placeholder="Search" autocomplete="off">
                                                            </span>
                                                            <span class="typeahead__button">
                                                                <button type="submit">
                                                                    <i class="typeahead__search-icon"></i>
                                                                </button>
                                                            </span>

                                                        </div>
                                                    </div>
                                                </form>
                                            </div>
                                            <div id="apple">
                                            </div>

                                            <div id="search_gene" class="hidden">
                                                <div class="input-group">
                                                    <input type="text" id="searchTerm" name="searchTerm" class="form-control" placeholder="">
                                                    <span class="input-group-btn">
                                                        <button type="submit" name="search" id="btnGo" class="btn btn-flat btn-default"><i class="fa fa-search"></i></button>
                                                    </span>
                                                </div>
                                                <div id="search_results_div"></div>
                                            </div>
                                        </div>
                                        <!-- /.tab-pane -->
                                        <div class="tab-pane" id="tab_2-2">
                                            <div id="search_history_div" style="width:100%">
                                            </div>
                                        </div>
                                        <!-- /.tab-pane -->
                                    </div>
                                    <!-- /.tab-content -->
                                </div>
                                <!-- /.box-body -->
                            </div>
                            <!-- /.box -->
                        </div>
                        <!-- .col -->
                    </div>
                    <div class="row">
                        <div class="col-md-12">
                            <div id="profile_plot_container_div" class="box box-default hidden">
                                <div class="box-header with-border">
                                    <h3 class="box-title">Profile Plot</h3>
                                    <div class="box-tools pull-right">
                                    </div>
                                </div>
                                <!-- /.box-header -->
                                <div class="box-body">
                                    <div class="row">
                                        <div class="col-sm-offset-1 col-sm-8">
                                            <div id="factor-view-select-wrapper">
                                                <select id="factor-view-select" multiple="multiple">
                                                </select>
                                            </div>
                                        </div>
                                        <div class="col-sm-2">
                                            <button id="btnFactPlot" type="button" class="btn btn-sm btn-flat btn-primary">
                                                <span class="fa fa-refresh" aria-hidden="true"></span> Update</button>
                                        </div>
                                    </div>
                                    <div class="row row-spacer">
                                    </div>
                                    <div class="row row-spacer">
                                    </div>
                                    <div class="row">
                                        <div class="col-sm-offset-1 col-sm-10">
                                            <div id="fact-svg-panel">
                                                <svg id="fact-svg-chart" width="100%" height="400px"></svg>
                                            </div>
                                        </div>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>


                <!--

                COLUMN 2

                //-->

                <div class="col-md-8">
                    <div class="row">
                        <div class="col-md-12">
                            <div id="gene_information_div" class="box box-default collapsed-box">
                                <div class="box-header with-border">
                                    <div id="div_current_gene_information_header">
                                        Please search for a term of interest.
                                    </div>
                                </div>
                                <div class="box-body">
                                    <div id="div_current_gene_information_body"></div>
                                </div>
                            </div>
                        </div>
                    </div>
                    <div class="row">
                        <div class="col-md-12">
                            <div id="lod_plot_container_div" class="box box-default hidden">
                                <div class="box-header with-border">
                                    <h3 class="box-title">LOD Plot</h3>
                                    <div class="box-tools pull-right">
                                    </div>
                                </div>
                                <!-- /.box-header -->
                                <div id="plot_lod_box_body" class="box-body">
                                    <div class="row">
                                        <div class="col-md-12">
                                            <strong>Scan Type</strong>
                                            <div class="btn-group" data-toggle="buttons">
                                                <label id="lbl_default" class="btn btn-xs btn-flat btn-default active">
                                                    <input type="radio" name="cond" id="cond_default" value="default" autocomplete="off" checked> Default
                                                </label>
                                                <label id="lbl_local" class="btn btn-xs btn-flat btn-default">
                                                    <input type="radio" name="cond" id="cond_local" value="local" autocomplete="off"> Local
                                                </label>
                                            </div>
                                        </div>
                                    </div>
                                    <div class="row">
                                        <div class="col-sm-12">
                                            <div id="plot_lod" style="width: 100%"></div>
                                            <!--
                                            Plotly.relayout('plot_lod', {'width':$("#plot_lod").width()});
                                            //-->
                                        </div>
                                    </div>

                                    <div class="row">
                                        <div class="col-sm-offset-8 col-sm-4">
                                            <input id="range_1" type="text" name="range_1" value="">
                                            Number of Data Points
                                        </div>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>

                    <div class="row">
                        <div class="col-md-12">
                            <div id="secondary_plot_container_div" class="nav-tabs-custom hidden">
                                <ul id="secondaryPlotTabs" class="nav nav-tabs">
                                    <li name="plot_effect" class="active"><a href="#tab_effect" data-toggle="tab">Effect Plot</a></li>
                                    <li name="plot_mediation" id="mediation_div"><a href="#tab_mediation" data-toggle="tab">Mediation Plot</a></li>
                                    <li name="plot_snpassoc"><a href="#tab_snpassoc" data-toggle="tab">SNP Association Plot</a></li>
                                    <li class="pull-right"><a href="#" class="text-muted">Click on the LOD plot to generate the selected graph</a></li>
                                </ul>
                                <div class="tab-content">
                                    <div class="tab-pane active" id="tab_effect">
                                        <div class="row row-spacer"></div>
                                        <div class="row">
                                            <div class="col-sm-offset-1 col-sm-11" id="plot_effect_settings">
                                                <strong>Settings</strong>
                                            </div>
                                        </div>
                                        <div class="row">
                                            <div class="col-sm-offset-1 col-sm-11">
                                                 Best Linear Unbiased Predictors (BLUPs)
                                                <input id="mediate_blup" name="mediate_blup" type="checkbox" data-size="mini">
                                            </div>
                                        </div>
                                        <div class="row">
                                            <div class="col-sm-12">
                                                <div id="plot_effect">
                                                </div>
                                            </div>
                                        </div>
                                        <div class="row">
                                            <div class="col-sm-12">
                                                <div id="plot_effect_2">
                                                </div>
                                            </div>
                                        </div>
                                    </div>
                                    <!-- /.tab-pane -->
                                    <div class="tab-pane" id="tab_mediation">
                                        <div id="plot_mediation" style="width:100%; height:100%">
                                        </div>
                                    </div>
                                    <!-- /.tab-pane -->
                                    <div class="tab-pane" id="tab_snpassoc">
                                        <div class="row">
                                            <div class="col-sm-offset-6 col-sm-4">
                                                <input id="rangeSNPWindow" type="text" name="rangeSNPWindow" value="1000000">
                                                SNP Window Size
                                            </div>
                                            <div class="col-sm-2">
                                                <button id="btnSNPAssocRefresh" type="button" class="btn btn-sm btn-flat btn-primary">
                                                    <span class="fa fa-refresh" aria-hidden="true"></span> Update</button>
                                            </div>
                                        </div>
                                        <div class="row">
                                            <div class="col-sm-12">
                                                <div id="plot_snpassoc" style="height:800px">
                                                </div>
                                            </div>
                                        </div>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        </section>
        <!-- /.content -->
    </div>
    <!-- /.content-wrapper -->

    <!-- Main Footer -->
    <footer class="main-footer">
        <!-- To the right -->
        <div class="pull-right hidden-xs">

        </div>
        <!-- Default to the left -->
        <strong>Created by <a href="http://churchill-lab.jax.org">The Churchill Lab</a>.</strong>
    </footer>

</div>
<!-- ./wrapper -->

<div id="question" style="display:none; cursor: default">
    <h1>Generating graph...</h1>
    <input type="button" id="stopGenerating" value="Stop" />
</div>

<!-- REQUIRED JS SCRIPTS -->

<!-- jQuery 2.2.3 -->
<script src="//code.jquery.com/jquery-2.2.4.min.js" integrity="sha256-BbhdlvQf/xTY9gja0Dq3HiwQF8LaCRTXxZKRutelT44=" crossorigin="anonymous"></script>

<!-- Bootstrap 3.3.7 -->
<script src="//maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>


<!-- AdminLTE App -->
<script src="{{g.URL_BASE}}/static/AdminLTE-2.3.7/dist/js/app.js"></script>
<!-- DataTables -->
<script src="{{g.URL_BASE}}/static/AdminLTE-2.3.7/plugins/datatables/jquery.dataTables.min.js"></script>
<script src="{{g.URL_BASE}}/static/AdminLTE-2.3.7/plugins/datatables/dataTables.bootstrap.min.js"></script>
<!-- SlimScroll -->
<script src="{{g.URL_BASE}}/static/AdminLTE-2.3.7/plugins/slimScroll/jquery.slimscroll.min.js"></script>
<!-- FastClick -->
<script src="{{g.URL_BASE}}/static/AdminLTE-2.3.7/plugins/fastclick/fastclick.js"></script>
<!-- ION Slider -->
<script src="{{g.URL_BASE}}/static/AdminLTE-2.3.7/plugins/ionslider/ion.rangeSlider.min.js"></script>

<!-- WHOZ API -->
<script src="//churchill-lab.jax.org/whoz/api/js/whoz_api.js"></script>

<!-- Highcharts //-->
<script src="//code.highcharts.com/stock/highstock.src.js"></script>
<script src="//code.highcharts.com/highcharts-more.js"></script>
<script src="//code.highcharts.com/modules/boost.src.js"></script>
<script src="{{g.URL_BASE}}/static/js/highcharts-downsample.max.js"></script>
<script src="//code.highcharts.com/modules/exporting.js"></script>
<script src="//code.highcharts.com/modules/canvas-tools.src.js"></script>
<script src="{{g.URL_BASE}}/static/highcharts-export/export-csv.js"></script>
<script src="{{g.URL_BASE}}/static/highcharts-export/jspdf.min.js"></script>
<script src="{{g.URL_BASE}}/static/highcharts-export/highcharts-export-clientside.js"></script>

<script src="{{g.URL_BASE}}/static/js/saveSvgAsPng.js"></script>

<script src="{{g.URL_BASE}}/static/js/jquery.blockUI.min.js"></script>
<script src="//rawgit.com/krunkosaurus/simg/v1.1.2/src/simg.js"></script>
<!--
If using d3.v4.js, queue is incluced, however d3.v4.js breaks factexpviewer.js
<script src="{{g.URL_BASE}}/static/js/d3.v3.js"></script>
//-->

<script src="//cdnjs.cloudflare.com/ajax/libs/d3/3.5.17/d3.min.js"></script>

<script src="{{g.URL_BASE}}/static/js/d3-queue.min.js"></script>
<script src="{{g.URL_BASE}}/static/bootstrap-switch/js/bootstrap-switch.min.js"></script>
<script src="{{g.URL_BASE}}/static/bootstrap-select/js/bootstrap-select.min.js"></script>
<script src="{{g.URL_BASE}}/static/bootstrap-notify-3.1.3/bootstrap-notify.min.js"></script>
<script src="{{g.URL_BASE}}/static/bootstrap-multiselect/js/bootstrap-multiselect.js"></script>
<script src="{{g.URL_BASE}}/static/js/seedrandom.js"></script>
<script src="{{g.URL_BASE}}/static/js/factexpviewer.js"></script>
<script src="{{g.URL_BASE}}/static/js/js.cookie.js"></script>
<script src="{{g.URL_BASE}}/static/jquery-typeahead-2.9.0/dist/jquery.typeahead.min.js"></script>
<script>
    <!--
    /*******************************************************************************************************************
     **
     ** GLOBAL VARIABLES
     **
     ******************************************************************************************************************/
    // GLOBAL VARIABLES
    var _G = {};

    // information on chromosomes
    _G.chromosomes = {'idx': [],
                      'nameToIdx': {},
                      'totalLength':-1};

    // gene ids
    /*
     gene_ids: {
         "ENSMUSG00000000001": {id:"ENSMUSG00000000001"},
         "ENSMUSG00000000008": {id:"ENSMUSG00000000008"}
     }
     */
    _G.genes = {gene_ids:{}};

    // protein ids and gene ids
    /*
     gene_ids: {
         "ENSMUSG00000000001": {
             id:"ENSMUSG00000000001", protein_ids:["ENSMUSP00000000001"]
         },
         "ENSMUSG00000000002": {
             id:"ENSMUSG00000000002", protein_ids:["ENSMUSP00000000049","ENSMUSP00000000050"]
         }
     },
     proteins_ids : [
         "ENSMUSP00000000001",
         "ENSMUSP00000000049",
         "ENSMUSP00000000090"
     ]
     */
    _G.proteins = {gene_ids:{}, proteins_ids: []};

    // gene id to search result map
    _G.searchResults = null;
    _G.searchResultsLevel = null;

    // APIS
    _G.api_root = '{{ CONF.API_R_BASE }}';
    _G.api_cores = '{{ CONF.API_CORES }}';
    _G.url_proxy = '{{ g.URL_BASE }}/proxy/';
    _G.url_proxy = '';
    _G.url_chromosomes = '{{ g.URL_BASE }}/static/chromosomes.mm.38.NO_Y.json';
    _G.max_lod_points = 0;

    _G.url_apis = {
        options: _G.api_root + '/options',
        ids: _G.api_root + '/ids',
        lod: _G.api_root + '/lodscan',
        mediate: _G.api_root + '/mediate',
        effect: _G.api_root + '/foundercoefs',
        expression: _G.api_root + '/expression',
        snpassoc: _G.api_root + '/snpassoc',
        phenotype: _G.api_root + '/phenotype',
        genes: 'http://churchill-lab.jax.org/whoz/api/id/',
        genesByLocation: 'http://churchill-lab.jax.org/whoz/api/search',
    };

    // search/click history
    _G.historyTable = null;

    // PLOTS

    _G.plots = {
        lod: {
            el: 'plot_lod',
            obj: null,
        },
        mediation: {
            el: 'plot_mediation',
            obj: null,
        },
        effect: {
            el: 'plot_effect',
            obj: null,
        },
        snpassoc: {
            el: 'plot_snpassoc',
            obj: null,
        },
        profile: {
            el: 'plot_profile',
            obj: null,
        }
    };

    _G.covar_factors = null;

    _G.current_gene = null;
    _G.current_gene_id = null;
    _G.current_protein_id = null;

    var WHOZ = null;

    var SPIN_GEARS = '<svg width="128px" height="128px" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100" preserveAspectRatio="xMidYMid" class="uil-gears"><rect x="0" y="0" width="100" height="100" fill="none" class="bk"></rect><g transform="translate(-20,-20)"><path d="M79.9,52.6C80,51.8,80,50.9,80,50s0-1.8-0.1-2.6l-5.1-0.4c-0.3-2.4-0.9-4.6-1.8-6.7l4.2-2.9c-0.7-1.6-1.6-3.1-2.6-4.5 L70,35c-1.4-1.9-3.1-3.5-4.9-4.9l2.2-4.6c-1.4-1-2.9-1.9-4.5-2.6L59.8,27c-2.1-0.9-4.4-1.5-6.7-1.8l-0.4-5.1C51.8,20,50.9,20,50,20 s-1.8,0-2.6,0.1l-0.4,5.1c-2.4,0.3-4.6,0.9-6.7,1.8l-2.9-4.1c-1.6,0.7-3.1,1.6-4.5,2.6l2.1,4.6c-1.9,1.4-3.5,3.1-5,4.9l-4.5-2.1 c-1,1.4-1.9,2.9-2.6,4.5l4.1,2.9c-0.9,2.1-1.5,4.4-1.8,6.8l-5,0.4C20,48.2,20,49.1,20,50s0,1.8,0.1,2.6l5,0.4 c0.3,2.4,0.9,4.7,1.8,6.8l-4.1,2.9c0.7,1.6,1.6,3.1,2.6,4.5l4.5-2.1c1.4,1.9,3.1,3.5,5,4.9l-2.1,4.6c1.4,1,2.9,1.9,4.5,2.6l2.9-4.1 c2.1,0.9,4.4,1.5,6.7,1.8l0.4,5.1C48.2,80,49.1,80,50,80s1.8,0,2.6-0.1l0.4-5.1c2.3-0.3,4.6-0.9,6.7-1.8l2.9,4.2 c1.6-0.7,3.1-1.6,4.5-2.6L65,69.9c1.9-1.4,3.5-3,4.9-4.9l4.6,2.2c1-1.4,1.9-2.9,2.6-4.5L73,59.8c0.9-2.1,1.5-4.4,1.8-6.7L79.9,52.6 z M50,65c-8.3,0-15-6.7-15-15c0-8.3,6.7-15,15-15s15,6.7,15,15C65,58.3,58.3,65,50,65z" fill="#000000"><animateTransform attributeName="transform" type="rotate" from="90 50 50" to="0 50 50" dur="1s" repeatCount="indefinite"></animateTransform></path></g><g transform="translate(20,20) rotate(15 50 50)"><path d="M79.9,52.6C80,51.8,80,50.9,80,50s0-1.8-0.1-2.6l-5.1-0.4c-0.3-2.4-0.9-4.6-1.8-6.7l4.2-2.9c-0.7-1.6-1.6-3.1-2.6-4.5 L70,35c-1.4-1.9-3.1-3.5-4.9-4.9l2.2-4.6c-1.4-1-2.9-1.9-4.5-2.6L59.8,27c-2.1-0.9-4.4-1.5-6.7-1.8l-0.4-5.1C51.8,20,50.9,20,50,20 s-1.8,0-2.6,0.1l-0.4,5.1c-2.4,0.3-4.6,0.9-6.7,1.8l-2.9-4.1c-1.6,0.7-3.1,1.6-4.5,2.6l2.1,4.6c-1.9,1.4-3.5,3.1-5,4.9l-4.5-2.1 c-1,1.4-1.9,2.9-2.6,4.5l4.1,2.9c-0.9,2.1-1.5,4.4-1.8,6.8l-5,0.4C20,48.2,20,49.1,20,50s0,1.8,0.1,2.6l5,0.4 c0.3,2.4,0.9,4.7,1.8,6.8l-4.1,2.9c0.7,1.6,1.6,3.1,2.6,4.5l4.5-2.1c1.4,1.9,3.1,3.5,5,4.9l-2.1,4.6c1.4,1,2.9,1.9,4.5,2.6l2.9-4.1 c2.1,0.9,4.4,1.5,6.7,1.8l0.4,5.1C48.2,80,49.1,80,50,80s1.8,0,2.6-0.1l0.4-5.1c2.3-0.3,4.6-0.9,6.7-1.8l2.9,4.2 c1.6-0.7,3.1-1.6,4.5-2.6L65,69.9c1.9-1.4,3.5-3,4.9-4.9l4.6,2.2c1-1.4,1.9-2.9,2.6-4.5L73,59.8c0.9-2.1,1.5-4.4,1.8-6.7L79.9,52.6 z M50,65c-8.3,0-15-6.7-15-15c0-8.3,6.7-15,15-15s15,6.7,15,15C65,58.3,58.3,65,50,65z" fill="#ffffff"><animateTransform attributeName="transform" type="rotate" from="0 50 50" to="90 50 50" dur="1s" repeatCount="indefinite"></animateTransform></path></g></svg>';
    var SPIN_INFINITY = '<svg width="128px" height="128px" viewBox="0 0 100 100" preserveAspectRatio="xMidYMid" xmlns:xlink="http://www.w3.org/1999/xlink" class="uil-inf"><rect x="0" y="0" width="100" height="100" fill="none" class="bk"></rect><path id="uil-inf-path" d="M24.3,30C11.4,30,5,43.3,5,50s6.4,20,19.3,20c19.3,0,32.1-40,51.4-40 C88.6,30,95,43.3,95,50s-6.4,20-19.3,20C56.4,70,43.6,30,24.3,30z" fill="none" stroke="#c0bb9c" stroke-width="1px" stroke-dasharray="5px"></path><circle cx="0" cy="0" r="5" fill="#ff0000"><animateMotion begin="0s" dur="1s" repeatCount="indefinite"><mpath xlink:href="#uil-inf-path"></mpath></animateMotion></circle><circle cx="0" cy="0" r="5" fill="#ff0000"><animateMotion begin="0.08s" dur="1s" repeatCount="indefinite"><mpath xlink:href="#uil-inf-path"></mpath></animateMotion></circle><circle cx="0" cy="0" r="5" fill="#ff0000"><animateMotion begin="0.16s" dur="1s" repeatCount="indefinite"><mpath xlink:href="#uil-inf-path"></mpath></animateMotion></circle><circle cx="0" cy="0" r="5" fill="#ff0000"><animateMotion begin="0.25s" dur="1s" repeatCount="indefinite"><mpath xlink:href="#uil-inf-path"></mpath></animateMotion></circle><circle cx="0" cy="0" r="5" fill="#ff0000"><animateMotion begin="0.33s" dur="1s" repeatCount="indefinite"><mpath xlink:href="#uil-inf-path"></mpath></animateMotion></circle></svg>';
    var SPIN_MAGNIFY = '<svg width="128px" height="128px" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 100 100" preserveAspectRatio="xMidYMid" class="uil-magnify"><rect x="0" y="0" width="100" height="100" fill="none" class="bk"></rect><g><circle fill="#8b8bfc" cx="47" cy="47" r="20" opacity="0.5"></circle><path d="M77.5,69.3l-6.2-6.2c-0.7-0.7-1.3-1.2-1.9-1.6c2.6-4,4.1-8.8,4.1-14c0-14.4-11.7-26.1-26.1-26.1S21.3,33.2,21.3,47.5 S33,73.6,47.4,73.6c5.4,0,10.4-1.6,14.5-4.4c0.5,0.7,1.1,1.4,1.9,2.2l5.8,5.8c2.9,2.9,7.1,3.5,9.2,1.3C81,76.4,80.4,72.2,77.5,69.3z M47.4,66.2c-10.3,0-18.7-8.4-18.7-18.6s8.4-18.6,18.7-18.6s18.7,8.4,18.7,18.6S57.7,66.2,47.4,66.2z" fill="#424242"></path><animateTransform attributeName="transform" type="translate" from="15 15" to="15 15" dur="1s" repeatCount="indefinite" values="15 15;-15 15;0 -10.98;15 15" keyTimes="0;0.33;0.66;1"></animateTransform></g></svg>';
    var BLOCK_STATE_SEARCHING = 1;
    var BLOCK_STATE_SEARCH_SELECTION = 2;
    var BLOCK_STATE_PROFILE_UPDATE = 3;
    var BLOCK_STATE_LOD_SELECTION = 4;
    var BLOCK_STATE_LOD_SCAN = 5;

    /*******************************************************************************************************************
     **
     ** GLOBAL FUNCTIONS
     **
     ******************************************************************************************************************/

    /**
     * Create a global exportCharts method.
     * @param {Array} charts - an array of charts
     * @param {Object} options - exporting options
     */
    Highcharts.exportCharts = function (charts, options) {
        options = Highcharts.merge(Highcharts.getOptions().exporting, options);

		// Get SVG asynchronously and then download the resulting SVG
        Highcharts.getSVG(charts, options, function (svg) {
            Highcharts.downloadSVGLocal(svg, options, function () {
                console.log("Failed to export on client side");
            });
        });
    };

    // Set global default options for all charts
    Highcharts.setOptions({
        exporting: {
            fallbackToExportServer: false // Ensure the export happens on the client side or not at all
        }
    });

    /**
     * Create a global getSVG method that takes an array of charts as an argument.
     * The SVG is returned as an argument in the callback.
     */
    Highcharts.getSVG = function (charts, options, callback) {
        let svgArr = [],
            top = 0,
            width = 0,
            svgResult = function (svgres) {
                // Grab width/height from exported chart
                let svgWidth = +svgres.match(
                        /^<svg[^>]*width\s*=\s*\"?(\d+)\"?[^>]*>/
                    )[1],
                    svgHeight = +svgres.match(
                        /^<svg[^>]*height\s*=\s*\"?(\d+)\"?[^>]*>/
                    )[1],
                    // Offset the position of this chart in the final SVG
                    svg = svgres.replace('<svg', '<g transform="translate(0,' + top + ')" ');
                svg = svg.replace('</svg>', '</g>');
                top += svgHeight;
                width = Math.max(width, svgWidth);
                svgArr.push(svg);
                if (svgArr.length === charts.length) {
                    return callback('<svg height="' + top + '" width="' + width + '" version="1.1" xmlns="http://www.w3.org/2000/svg">' + svgArr.join('') + '</svg>');
                }
            };
        for (let i = 0; i < charts.length; ++i) {
            console.log(charts[i]);
            charts[i].getSVGForLocalExport(options, {}, function () {
                console.log("Failed to get SVG");
            }, svgResult);
        }
    };

    // Extended disable function
    jQuery.fn.extend({
        disable: function(state) {
            return this.each(function() {
                let $this = $(this);
                if($this.is('input, button, textarea, select'))
                    this.disabled = state;
                else
                    $this.toggleClass('disabled', state);
            });
        }
    });

    /*******************************************************************************************************************
     **
     ** GENERAL FUNCTIONS
     **
     ******************************************************************************************************************/

    /**
     * Block user input.
     * @param {string} state - the current state of selection
     */
    function blockUser(state) {
        if (state === BLOCK_STATE_SEARCHING) {
            $('#search_box_div').block({message:SPIN_MAGNIFY, overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#gene_information_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#profile_plot_container_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#lod_plot_container_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#secondary_plot_container_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
        } else if (state === BLOCK_STATE_SEARCH_SELECTION) {
            $('#search_box_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#gene_information_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#profile_plot_container_div').block({message:SPIN_GEARS, overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#lod_plot_container_div').block({message:SPIN_GEARS, overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#secondary_plot_container_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
        } else if (state === BLOCK_STATE_PROFILE_UPDATE) {
            $('#search_box_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#gene_information_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#profile_plot_container_div').block({message:SPIN_GEARS, overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#lod_plot_container_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#secondary_plot_container_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
        } else if (state === BLOCK_STATE_LOD_SELECTION) {
            $('#search_box_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#gene_information_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#profile_plot_container_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#lod_plot_container_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#secondary_plot_container_div').block({message:SPIN_GEARS, overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
        } else if (state === BLOCK_STATE_LOD_SCAN) {
            $('#search_box_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#gene_information_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#profile_plot_container_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#lod_plot_container_div').block({message:SPIN_GEARS, overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
            $('#secondary_plot_container_div').block({message:'', overlayCSS: {opacity: 0.2}, css:{backgroundColor:'', border:'none'}});
        }

    }

    /**
     * Unblock user input.
     */
    function unblockUser() {
        $('#search_box_div').unblock();
        $('#gene_information_div').unblock();
        $('#profile_plot_container_div').unblock();
        $('#lod_plot_container_div').unblock();
        $('#secondary_plot_container_div').unblock();
    }

    /**
     * Reset all the secondary plots to not be displayed.
     */
    function resetSecondaryPlots() {
        $('#plot_mediation').html('');
        $('#plot_effect').html('');
        $('#plot_snpassoc').html('');
    }

    /**
     * Format Mbp to two decimal places.
     * @param {number} Mbp - the number to round
     * @return {string} Mbp formatted to two decimal places
     */
    function formatMbp(Mbp) {
        return Number(Mbp).toFixed(2);
    }

    /**
     * Simple compare function.
     * @param {object} a - first object
     * @param {number} a.x - first number
     * @param {object} b - second object
     * @param {number} b.x - second number
     * @return {number} 0 if equal, 1 if a.x > b.x, -1 if a.x < b.x
     */
    function compareX(a, b) {
        if (a.x  < b.x)
            return -1;
        if (a.x > b.x)
            return 1;
        return 0;
    }

    /**
     * Move array element from fromIndex to toIndex.
     * @param {Array} arr - the array
     * @param {number} fromIndex - the from index
     * @param {number} toIndex - the to index
     */
    function arraymove(arr, fromIndex, toIndex) {
        let element = arr[fromIndex];
        arr.splice(fromIndex, 1);
        arr.splice(toIndex, 0, element);
    }

    /**
     * Display an error to the user.
     * @param {string} title - the title of the message
     * @param {string} message - the message
     */
    function displayError(title, message) {
        $.notify({
                title: '<h5>' + title + '</h5>',
                message: message
            }, {
                type: 'danger'
            });

        // make sure all waiting images are no longer
        $.each($("div[id$='wait']"), function(index, element) {
            $(element).html('');
        });
    }

    /**
     * Check if cookies are enabled.
     * @returns {boolean} whether cookies are enables or not
     */
    function cookiesEnabled() {
        let isValid = Cookies.set('checkme', 'valid', {expires: 1}) && Cookies.get('checkme') === 'valid';
        Cookies.remove('checkme');
        return isValid;
    }

    /**
     * Keep track of gene/protein click history.
     * @param {String} gene_id - the gene identifier
     * @param {String} protein_id - the protein identifier
     * @param {String} symbol - the gene symbol
     * @param {String} level - the level ('mrna' or 'protein')
     */
    function addHistory(gene_id, protein_id, symbol, level) {
        let unique_id = gene_id;

        if (protein_id) {
            unique_id += "," + protein_id;
        }

        // cookies, everywhere cookie
        let history_data = Cookies.getJSON('history_data');
        let index = -1;

        // build the history of 'clicked' genes
        if (history_data) {
            // move to the top if already here
            for (let i in history_data) {
                if (history_data[i]['id'] === unique_id) {
                    index = i;
                    break;
                }
            }
            if (index > 0) {
                arraymove(history_data, index, 0);
            } else if (index === -1) {
                let history_info = {'id': unique_id, 'g': gene_id, 'p': protein_id, 's': symbol, 'l':level};
                history_data.unshift(history_info);

                if (history_data.length > 20) {
                    history_data.pop();
                }
            } else {
                console.log('doing nothing, returning');
            }
        } else {
            history_data = [{'id': unique_id, 'g': gene_id, 'p': protein_id, 's': symbol, 'l':level}];
        }

        // attempt to set the cookie, if there is to much data the cookie won't set,
        // so we need to check and adjust the array size
        Cookies.set("history_data", history_data);
    }

    /**
     * Build history search table.
     */
    function buildHistory() {
        // cookies, everywhere cookie
        if (!cookiesEnabled()) {
            $('#search_history_div').html("<br><span class='text-danger'>History requires that cookies are enabled in your browser.</span>");
            return;
        }

        let history_data = Cookies.getJSON("history_data");

        if (history_data) {
            let historyTableData = [];

            for (let _h in history_data) {
                let h = history_data[_h];
                let d = {
                    'gene_id': h['g'],
                    'symbol': h['s'],
                    'protein_id': h['p'],
                    'level': h['l']
                };

                historyTableData.push(d);
            }

            if (history_data.length === 1) {
                $('#searchHistoryStatus').html("<p class='text-right text-muted'><em><small>Showing last gene viewed</small></em></p>");
            } else {
                $('#searchHistoryStatus').html("<p class='text-right text-muted'><em><small>Showing last " + history_data.length.toLocaleString() + " genes viewed</small></em></p>");
            }

            let tbl = '<table id="search_history_table" class="table table-bordered table-hover table-condensed"></table>';
            $('#search_history_div').html(tbl);

            let historyTableColumns = [];

            historyTableColumns.push({
                title:'ID',
                data:"gene_id",
                render: function(data, type, row, meta) {
                    if (row['level'] === 'mrna') {
                        if (type === 'display') {
                            if (_G.genes.gene_ids[data]) {
                                return '<div style="white-space: nowrap"><a href="#" level="' + row['level'] + '" gene_id="' + data + '">' + data + '</a></div>';
                            }

                            return '<div style="white-space: nowrap">' + data + '</div>';
                        }

                        return data;
                    } else {
                        if (type === 'display') {
                            let ps = [];
                            if (_G.proteins.gene_ids[data]) {
                                let ret = '<div style="white-space: nowrap">' + data;
                                ret += '</div><div style="padding-left: 20px;">';
                                ret += '<em>';

                                for (p in _G.proteins.gene_ids[data].protein_ids) {
                                    ps.push('<a href="#" level="' + row['level'] + '" gene_id="' + data + '" protein_id="' + _G.proteins.gene_ids[data].protein_ids[p] + '">' + _G.proteins.gene_ids[data].protein_ids[p] + '</a>');
                                }

                                return ret + ps.join('<br>') + "</em></div></small>";

                            }

                            return '<div style="white-space: nowrap">' + data + '</div>';
                        }

                        return data;
                    }
                }
            });

            historyTableColumns.push({title:'Symbol', data:"symbol"});

            _G.historyTable = $("#search_history_table").DataTable({
                data: historyTableData,
                paging: false,
                order: [],   // prevent automatic ordering
                scrollY: 200,
                searching: false,
                columns: historyTableColumns,
                autoWidth:true,
                responsive:true,
            });


            $("#search_history_table a").on("click", function (e) {
                e.preventDefault();

                let _this = $(this);
                let regressLocal = $("#cond_local").is(":checked");

                selectGeneProtein(_this.attr('gene_id'), _this.attr('protein_id'), _this.attr('level'), regressLocal);

                return false;
            });

        } else {
            $('#search_history_div').html("<br><span class='text-danger'>No results found.</span>");
        }
    }

    /**
     * Get the selected factors.
     * @returns {Array} an array of strings
     */
    function getFactorOrder() {
        let selected = [];
        $('#factor-view-select option:selected').each(function() {
            selected.push([$(this).val(), $(this).attr('order')]);
        });

        selected.sort(function(a, b) {
            return a[1] - b[1];
        });

        let ordered = [];
        for (let i = 0; i < selected.length; i++) {
            ordered.push(selected[i][0]);
        }

        return ordered;
    }

    /*******************************************************************************************************************
     **
     ** LAYOUT ALGORITHMS
     **
     ******************************************************************************************************************/

    /**
     * If the slot has space or not
     * @param {Object} feature - the feature to check
     * @param {number} feature.position_start - the strat position
     * @param {number} feature.end - the end position
     * @param {Array} features_in_slot - an array of features in this slot
     * @return {boolean} true if slot has space for the feature
     */
    function slotHasSpace(feature, features_in_slot) {
        if (!features_in_slot) {
            return true;
        }

        for (let i = 0; i < features_in_slot.length; i++) {
            let subject = features_in_slot[i];

            if (((feature.position_start <= subject.position_start) && (feature.position_end >= subject.position_start)) ||
                ((feature.position_start >= subject.position_start) && (feature.position_start <= subject.position_end))) {
                return false;
            }
        }

        return true;
    }

    /**
     * Sort the features by the slot.
     * @param {Array} features - an array of features in this slot
     * @return {Array} the sorted slots
     */
    function sortFeaturesBySlot(features) {
        let slots = [];
        for (let i = 0; i < features.length; i++) {
            let feature = features[i];
            if (!slots[feature.slot]) {
                slots[feature.slot] = [];
            }
            slots[feature.slot].push(feature);
        }
        return slots;
    }

    /**
     * Determine where each feature should be located.
     * @param {Array} features - an array of features
     */
    function layoutFeatures(features) {
        let allocated = [];
        let remaining = features;
        let needed_slots = 0;

        for (let i = 0; i < remaining.length; i++) {
            let features_by_slot = sortFeaturesBySlot(allocated);
            let current = remaining[i];
            let slot = 0;
            while (true) {
                if (slotHasSpace(current, features_by_slot[slot])) {
                    current.slot = slot;
                    allocated.push(current);
                    if (slot > needed_slots) {
                        needed_slots = slot;
                    }
                    break;
                }
                slot++;
            }
        }

    }

    /*******************************************************************************************************************
     **
     ** DOWNLOAD FUNCTIONS
     **
     ******************************************************************************************************************/

    /**
     * Download data.
     * @param {string} URL - the URL to download data
     * @param {string} description - descriptive text
     * @param {function} callback - the callback function
     */
    function downloadData(URL, description, callback) {
        /*
         From d3.js documentation:

         When a task completes, it must call the provided callback.

         The first argument to the callback should be null if the task is successful,
         or the error if the task failed.

         The optional second argument to the callback is the return value of the task.
         (To return multiple values from a single callback, wrap the results in an object or array.)
         */
        URL = _G.url_proxy + URL;

        console.log('Downloading: ', URL);

        $.ajax({
            url: URL,
            method: 'GET'
        }).done(function(data, textStatus, jqXHR) {
            console.log('Download of ' + description + ':', data);
            callback(null, data);
        }).fail(function(jqXHR, textStatus, errorThrown) {
            displayError('Server Error', 'Unable to download ' + description);
            console.error(description, textStatus);
            callback(textStatus, null);
        });
    }

    /**
     * Download options for the viewer.
     * @param {function} callback - the function to call upon completion
     */
    function downloadOptions(callback) {
        let URL = _G.url_apis.options;
        downloadData(URL, 'OPTIONS', callback);
    }

    /**
     * Download chromosomes for the viewer.
     * @param {function} callback - the function to call upon completion
     */
    function downloadChromosomes(callback) {
        let URL = _G.url_chromosomes;
        downloadData(URL, 'CHROMOSOMES', callback);
    }

    /**
     * Download the genes for the viewer.
     * @param {function} callback - the function to call upon completion
     */
    function downloadGenes(callback) {
        let URL = _G.url_apis.ids + '/mrna';
        downloadData(URL, 'GENES', callback);
    }

    /**
     * Download the proteins for the viewer.
     * @param {function} callback - the function to call upon completion
     */
    function downloadProteins(callback) {
        let URL = _G.url_apis.ids + '/protein';
        downloadData(URL, 'PROTEINS', callback);
    }

    /**
     * Download the phenotypes for the viewer.
     * @param {function} callback - the function to call upon completion
     */
    function downloadPhenotypes(callback) {
        let URL = _G.url_apis.ids + '/phenotype';
        downloadData(URL, 'PHENOTYPES', callback);
    }

    /**
     * Download LOD scan data from the server.
     * @param {string} id - unique identifier
     * @param {string} level - current level
     * @param {boolean} regressLocal - whether to regress locally or not
     * @param {function} callback - callback function to execute
     */
    function downloadLODData(id, level, regressLocal, callback) {
        let URL = _G.url_apis.lod;
        URL += '?id=' + id;
        URL += '&level=' + level;
        URL += '&regress_local=' + regressLocal;
        URL += '&ncores=' + _G.api_cores;
        downloadData(URL, 'LOD DATA', callback);
    }

    /**
     * Download SNP association data from the server.
     * @param {string} id - unique identifier
     * @param {string} level - current level
     * @param {string} chromosome - chromosome
     * @param {string} location - location on the chromosome
     * @param {number} windowSize - size of the window on each side of location
     * @param {function} callback - callback function to execute
     */
    function downloadSNPAssocData(id, level, chromosome, location, windowSize, callback) {
        let URL = _G.url_apis.snpassoc;
        URL += '?id=' + id;
        URL += '&level=' + level;
        URL += '&chrom=' + chromosome;
        URL += '&location=' + location;
        URL += '&window_size=' + windowSize;
        URL += '&ncores=' + _G.api_cores;
        downloadData(URL, 'SNP ASSOC DATA', callback);
    }

    /**
     * Download mediation data from the server.
     * @param {string} id - unique identifier
     * @param {string} marker_id - unique marker identifier
     * @param {string} level - current level
     * @param {function} callback - callback function to execute
     */
    function downloadMediationData(id, marker_id, level, callback) {
        let URL = _G.url_apis.mediate;
        URL += '?id=' + id;
        URL += '&mid=' + marker_id;
        URL += '&level=' + level;
        downloadData(URL, 'MEDIATION DATA', callback);
    }

    /**
     * Download effect data from the server.
     * @param {string} id - unique identifier
     * @param {string} chromosome - chromosome
     * @param {string} level - current level
     * @param {boolean} blup - whether to do blup analysis or not
     * @param {boolean} regressLocal - whether to regress locally or not
     * @param {function} callback - callback function to execute
     */
    function downloadEffectData(id, chromosome, level, blup, regressLocal, callback) {
        let URL = _G.url_apis.effect;
        URL += '?id=' + id;
        URL += '&chrom=' + chromosome;
        URL += '&level=' + level;
        URL += '&blup=' + blup;
        URL += '&regress_local=' + regressLocal;
        URL += '&ncores=' + _G.api_cores;
        downloadData(URL, 'EFFECT DATA', callback);
    }

    /**
     * Download mediation data from the server.
     * @param {string} id - unique identifier
     * @param {string} level - current level
     * @param {function} callback - callback function to execute
     */
    function downloadProfileData(id, level, callback) {
        let URL = _G.url_apis.expression;
        URL += '?id=' + id;
        URL += '&level=' + level;
        downloadData(URL, 'PROFILE DATA', callback);
    }

    /**
     * Download gene data from the server.
     * @param {string} id - unique identifier
     * @param {function} callback - callback function to execute
     */
    function downloadGeneData(id, callback) {
        let URL = _G.url_apis.genes + id;
        downloadData(URL, 'GENE DATA', callback);
    }

    /**
     * Download gene data from the server by location.
     * @param {string} chromosome - chromosome
     * @param {number} start - start position
     * @param {number} end - end position
     * @param {function} callback - callback function to execute
     */
    function downloadGenesByLocationData(chromosome, start, end, callback) {
        let location = chromosome + ':' + start + '-' + end;
        let URL = _G.url_apis.genesByLocation;
        URL += '?term=' + location;
        downloadData(URL, 'GENES BY LOCATION', callback);
    }

    /**
     * Download phenotype data from the server.
     * @param {string} id - unique identifier
     * @param {function} callback - callback function to execute
     */
    function downloadPhenotypeData(id, callback) {
        let URL = _G.url_apis.phenotype;
        URL += '?id='+ id;
        downloadData(URL, 'PHENOTYPE DATA', callback);
    }

    /*******************************************************************************************************************
     **
     ** PLOT FUNCTIONS
     **
     ******************************************************************************************************************/

    /**
     * Plot LOD data.
     *
     * Data returned from API is:
     *
     *     {"data":[{"1_40055","1",40055,2.4787},
     *              {"1_87283","1",87283,2.4787},
     *              ...
     *              {"1_134512","1",134512,:2.4787}
     *             ]}
     *
     * @param {Object} lodData - the data from the LOD scan
     * @param {string} id - the unique identifier
     * @param {string} level - the current level
     */
    function plotLOD(lodData, id, level) {
        let dataByChrom = {};
        let maxLOD = -Infinity;
        let minLOD = Infinity;
        let xAxisTitle = '';
        let newLodData = [];

        if (lodData !== null) {
            maxLOD = lodData[0][3];
            minLOD = -2;

            if ((level === 'mrna') || (level === 'protein')) {
                xAxisTitle = id + ' (' + _G.current_gene.symbol + ')';
            } else {
                // pheno
                xAxisTitle = id;
                if (id !== _G.phenotypes[id].short_name) {
                    xAxisTitle += (' (' + _G.phenotypes[id].short_name + ')');
                }
            }

            $.each(lodData, function(index, element) {
                //
                // element = {"[id]","[chr]",[position],[data]}
                //
                if (element[1] in _G.chromosomes.nameToIdx) {
                    let chr = _G.chromosomes.idx[_G.chromosomes.nameToIdx[element[1]]];
                    let x = chr.start + (element[2]);
                    let y = element[3];

                    //
                    // chr = {chromosome: "Y", end:2725521390, length: 91744698,
                    //        mid: 2679649040, name: "Y", order: 21, start: 2633776692 }
                    //
                    if (!(chr.chromosome in dataByChrom)) {
                        dataByChrom[chr.chromosome] = [];
                    }

                    let d = {
                        x: x,
                        y: y,
                        id: element[0],
                        text: element[3],
                        chr: chr.chromosome,
                        chrPos: element[2],
                    };

                    newLodData.push(d);
                    dataByChrom[chr.chromosome].push(d);

                    if (element[3] > maxLOD) {
                        maxLOD = element[3];
                    }
                }
            });

            maxLOD += 5;
        }

        newLodData.sort(compareX);

        _G.lod_traces = {};
        _G.lodDataByChromosome = dataByChrom;

        // create scatter for each chromosome
        let series = [];
        $.each(dataByChrom, function(key, value) {
            let s = {
                type: 'scatter',
                x: value['x'],
                y: value['y'],
                text: value['id'],
                name: key,
                hoverinfo: 'text',
                mode: 'lines',
                opacity: 100.0,
                line: {
                    color: 'rgb(0,0,0)',
                    width: 1
                }
            };

            _G.lod_traces[key] = s;
            series.push(s);
        });


        //
        // build the vertical chromosome bars
        //
        let chromosomeAxisVals = [];
        let chromosomeAxisText = {};
        let chromosomeBands = [];

        $.each(_G.chromosomes.idx, function(index, element) {

            chromosomeAxisVals.push(element.mid);
            chromosomeAxisText[element.mid] = element.name;

            // TODO: style this
            let fillColor = '#ffffff';
            if ((index % 2) === 1) {
                chromosomeBands.push({
                    color: fillColor,
                    from: element.start,
                    to: element.end
                });
            }
        });

        let plotShapes = [];
        let plotAnnotations = [];

        {% if CONF.PLOT_LOD_LINES %}
        //
        // build the horizontal lines
        //
        let lineVals = [
            {% for l in CONF.PLOT_LOD_LINES %}
                {'val':{{l.val}}, 'color':'{{l.color}}', 'text':'{{l.text }}' },
            {% endfor %}
        ];
        let _plotLines = [
            {% for l in CONF.PLOT_LOD_LINES %}
                {value:{{l.val}}, color:'{{l.color}}', zIndex:5, width:2, label:{
                    text:'{{l.text }}', align:'right'
                } },
            {% endfor %}
            ];
        for (let l in lineVals) {
            plotShapes.push({
                type: 'line',
                xref: 'paper',
                x0: 0,
                y0: lineVals[l].val,
                x1: 1,
                y1: lineVals[l].val,
                line: {
                    color: lineVals[l].color,
                    width: 1.5
                },
                layer: 'above',
            });

            plotAnnotations.push({
                x: 1,
                y: lineVals[l].val,
                xref: 'paper',
                yref: 'y',
                text: lineVals[l].text,
                xanchor: 'left',
                yanchor: 'middle',
                showarrow: false,
            });
        }
        {% endif %}

        /*
        TODO: make sure other chromosome data files are working
        console.log('chromosomeAxisVals=', chromosomeAxisVals);
        console.log('chromosomeAxisText=', chromosomeAxisText);
        */

        _G.newLodData = newLodData;
        _G.lodHover = null;

        _G.plots.lod.obj = Highcharts.chart({
            title: {
                text: xAxisTitle
            },

            chart: {
                alignTicks: false,
                renderTo: _G.plots.lod.el,
                zoomType: 'x',
                events: {
                    click: function(event) {
                        let plot=$('#secondaryPlotTabs').find('.active').attr('name');
                        console.log('chart event', event);
                        console.log('chart this', this);
                        console.log('hover=', _G.lodHover);

                        if (event.target.tagName === 'tspan') {
                            return;
                        }

                        // TODO: is _G.lodHover a good idea or can we get it from event

                        if (plot === 'plot_mediation') {
                            generateMediationPlot(id, _G.lodHover.id, level);
                        } else if (plot === 'plot_effect') {
                            generateEffectPlot(id, level, _G.lodHover.chr, $("#mediate_blup").is(':checked'), $("#cond_local").is(":checked"));
                        } else if (plot === 'plot_snpassoc') {
                            generateSNPAssocPlot(id, level, _G.lodHover.chr, _G.lodHover.chrPos, _G.sliderSNP.result.from_value);
                        } else {
                            console.error('Unknown plot ', plot);
                        }
                    }
                }
            },

            xAxis: {
                opposite: true,
                tickWidth: 0,
                labels: {
                    formatter: function () {
                        return chromosomeAxisText[this.value];
                    }
                },
                tickPositions: chromosomeAxisVals,
                plotBands: chromosomeBands,
            },

            yAxis: {
                title: {
                    text: 'LOD'
                },
                plotLines: _plotLines
            },

            tooltip: {
                positioner: function(labelWidth, labelHeight, point) {
                    let tooltipX, tooltipY;
                    let chart = _G.plots.lod.obj;

                    if (point.plotX + labelWidth > chart.plotWidth) {
                        tooltipX = point.plotX + chart.plotLeft - labelWidth - 20;
                    } else {
                        tooltipX = point.plotX + chart.plotLeft + 20;
                    }

                    if (point.plotY + labelHeight > chart.plotHeight) {
                        tooltipY = point.plotY + chart.plotTop - labelHeight - 20;
                    } else {
                        tooltipY = point.plotY + chart.plotTop - 20;
                    }

                    return {
                        x: tooltipX,
                        y: tooltipY
                    };


                },
                useHTML: true,
                formatter: function () {
                    _G.lodHover = this.point;
                    return '<b>' + this.point.id + '</b><br/>LOD: ' + this.point.y + '<br/>Chromosome: ' + this.point.chr + '<br/>Position: ' + this.point.chrPos;
                }
            },

            plotOptions: {
                series: {
                    cursor: 'pointer',
                    events: {
                        click: function(event) {
                            //event.point holds a pointer to the nearest point on the graph.
                            // The this keyword refers to the Series object.
                            // Log to console
                            let plot=$('#secondaryPlotTabs').find('.active').attr('name');

                            if (plot === 'plot_mediation') {
                                generateMediationPlot(id, event.point.id, level);
                            } else if (plot === 'plot_effect') {
                                generateEffectPlot(id, level, event.point.chr, $("#mediate_blup").is(':checked'), $("#cond_local").is(":checked"));
                            } else if (plot === 'plot_snpassoc') {
                                generateSNPAssocPlot(id, level, event.point.chr, event.point.chrPos, _G.sliderSNP.result.from_value);
                            } else {
                                console.error('Unknown plot ', plot);
                            }
                        }
                    }
                }
            },

            series: [{
                type: 'line',
                turboThreshold: 0,
                showInLegend: false,
                downsample: {
                    threshold: 10000 // 0 disables downsampling
                },
                data: newLodData,
                lineWidth: 0.5,
                color: 'black', // Color value
            }],

            credits: {
                enabled: false // A plugin for highcharts is of course using highcharts
            },

            exporting: {
                buttons: {
                    contextButton: {
                        menuItems: [{
                            text: 'Download JPEG image',
                            onclick: function () {
                                this.exportChartLocal({
                                    filename: id + '_LOD',
                                    type: Highcharts.exporting.MIME_TYPES.JPEG
                                });
                            }
                        }, {
                            text: 'Download PNG image',
                            onclick: function () {
                                this.exportChartLocal({
                                    filename: id + '_LOD',
                                    type: Highcharts.exporting.MIME_TYPES.PNG
                                });
                            },
                            separator: false
                        }, {
                            text: 'Download PDF document',
                            onclick: function () {
                                this.exportChartLocal({
                                    filename: id + '_LOD',
                                    type: Highcharts.exporting.MIME_TYPES.PDF
                                });
                            },
                            separator: false
                        }]
                    }
                }
            }
        });
    }

    /**
     * Plot the SNP Association plot.
     * @param {Array} snpData - array of SNP information
     * @param {Array} geneData - array of gene objects
     */
    function plotSNPAssoc(snpData, geneData) {
        let dataByChrom = {};
        let maxLODScore = null;
        let xAxisTitle = '';

        let _title = '';

        let currentLevel = getCurrentLevel();

        if (currentLevel === 'mrna') {
            _title = _G.current_gene_id + " (" + _G.current_gene.symbol + ")";
        } else if (currentLevel === 'protein') {
            _title = _G.current_protein_id + " (" + _G.current_gene.symbol + ")";
        } else {
            _title = _G.current_phenotype_id;
            if (_G.current_phenotype_id !== _G.phenotypes[_G.current_phenotype_id].short_name) {
                _title += (' (' + _G.phenotypes[_G.current_phenotype_id].short_name + ')');
            }
        }

        // this chart will be inverted so we will have to "rotate/manipulate" the data

        // sort the SNP data
        snpData.sort(function(a, b) {
            return a.pos - b.pos;
        });

        let snps = [];

        if (snpData) {
            maxLODScore = snpData[0].lod;
        }

        $.each(snpData, function(index, element) {
            if (element.lod > maxLODScore) {
                maxLODScore = element.lod;
            }

            snps.push({
                id: element.snp,
                alleles: element.alleles,
                lod: element.lod,
                // NOTE: since chart is inverted we are manipulating
                x: -1.0 * element.lod,
                y: element.pos * 1000000,
                csq: element.csq
            })

        });

        maxLODScore += 1;

        let series = [];

        series.push({
            name: 'seriesSNPs',
            data: snps,
            turboThreshold: 0,
            color: 'rgba(223, 83, 83, .5)',
            xAxis: 'snpTop',
            type: 'scatter'
        });

        let genes = [];
        $.each(geneData.result.matches, function(index, value) {
            if (value.symbol !== null) {
                genes.push(value);
            }
        });

        // generate the gene layout
        genes.sort(function(a, b) {
            return a.position_start - b.position_start;
        });
        layoutFeatures(genes);
        genes.sort(function(a, b) {
            return a.position_start - b.position_start;
        });

        let categories = [];
        let data = [];

        $.each(genes, function(index, value) {
            if (!(value.slot in categories)) {
                categories.push(value.slot);
            }
            let e = value;
            e.x = value.slot;
            e.low = value.position_start;
            e.high = value.position_end;
            data.push(e);
        });

        series.push({
            name: 'seriesGenes',
            data: data,
            xAxis: 'genesBottom',
            type: 'columnrange',
        });

        _G.plots.snpassoc.obj = Highcharts.chart({
            chart: {
                renderTo: _G.plots.snpassoc.el,
                inverted: true,
                panning: true,
                panKey: 'shift',
                zoomType: 'y',
                animation: false,
            },

            title: {
                text: _title,
            },

            yAxis: [{
                labels: {
                    enabled: true
                },
                title: {
                    text: ''
                },
                scrollbar: {
                    enabled: true
                },
                crosshair: {
                    snap: false
                }
            }],

            xAxis: [
                {
                    id: 'snpTop',
                    title: {
                        text: 'LOD'
                    },
                    max:0,
                    min: -1.0 * maxLODScore,
                    offset: 0,
                    height: '50%',
                    labels: {
                        formatter: function () {
                            return -1 * this.value;
                        }
                    }
                },
                {
                    id: 'genesBottom',
                    min: 0,
                    max: 6,
                    categories: categories,
                    title: {
                        text: ""
                    },
                    labels: {
                        enabled: false
                    },
                    // lineWidth: 0,
                    minorGridLineWidth: 0,
                    // lineColor: 'transparent',
                    minorTickLength: 0,
                    tickLength: 0,
                    offset: 0,
                    top: '55%',
                    height: '45%'
                }
            ],

            tooltip: {
                 positioner: function(labelWidth, labelHeight, point){
                    let tooltipX, tooltipY;
                    let chart = _G.plots.snpassoc.obj;
                    if (point.plotX + labelWidth > chart.plotWidth) {
                        tooltipX = point.plotX + chart.plotLeft - labelWidth - 20;
                    } else {
                        tooltipX = point.plotX + chart.plotLeft + 20;
                    }
                    tooltipY = point.plotY + chart.plotTop - 20;
                    return {
                        x: tooltipX,
                        y: tooltipY
                    };
                },
                useHTML: true,

                shared: false,
                followPointer: true,
                formatter: function () {
                    let text = "";
                    if (this.series.name === 'seriesSNPs') {
                        let csq = this.point.csq.split(',');
                        let csqs = csq.join('<br/>');

                        text = '<b>' + this.point.id + '</b><br/>' +
                            'LOD: ' + this.point.lod + '<br/>' +
                            'Position: ' + formatMbp(this.point.y / 100000) + '<br/>' +
                            'Alleles: ' + this.point.alleles + '<br/>' +
                            'CSQ: ' + csqs;
                    } else {
                        text = '<b>' + this.point.symbol + '</b><br/>' +
                            'Ensembl ID: ' + this.point.ensembl_gene_id + '<br/>' +
                            'Name: ' + this.point.name + '<br/>' +
                            'Position: ' + this.point.match_value + ' (' + this.point.strand + ')';

                    }
                    return text;
                }
            },

            plotOptions: {
                scatter: {
                    marker: {
                        radius: 2,
                        states: {
                            hover: {
                                enabled: false,
                                lineColor: 'rgb(100,100,100)'
                            }
                        }
                    },
                },
                columnrange: {
                    grouping: false,

                    dataLabels: {
                        //yLow:15,
                        enabled: true,
                        inside: true,
                        overlap:true,
                        align: 'center',
                        formatter: function() {
                            return this.point.symbol;
                        }
                    }
                }
            },
            legend: {
                enabled: false
            },

            series: series,

            credits: {
                enabled: false // A plugin for highcharts is of course using highcharts
            },
            exporting: {
                buttons: {
                    contextButton: {
                        menuItems: [{
                            text: 'Download JPEG image',
                            onclick: function () {
                                this.exportChartLocal({
                                    filename: _title + "_SNP",
                                    type: Highcharts.exporting.MIME_TYPES.JPEG
                                });
                            }
                        }, {
                            text: 'Download PNG image',
                            onclick: function () {
                                this.exportChartLocal({
                                    filename: _title + "_SNP",
                                    type: Highcharts.exporting.MIME_TYPES.PNG
                                });
                            },
                            separator: false
                        }, {
                            text: 'Download PDF document',
                            onclick: function () {
                                this.exportChartLocal({
                                    filename: _title + "_SNP",
                                    type: Highcharts.exporting.MIME_TYPES.PDF
                                });
                            },
                            separator: false
                        }]
                    }
                }
            }
        });
    }

    /**
     * Plot the effect data.
     * @param {Object} effectData - the effect data
     * @param {string} chromosome - the chromosome
     */
    function plotEffect(effectData, chromosome) {
        console.log('Plotting effect data');

        let strainInfo = {};
        {% for s in CONF.PLOT_EFFECT_STRAINS %}
            strainInfo["{{ s['key'] }}"] = {
                'color':"{{ s['color'] }}",
                'name':"{{ s['name'] }}",
                'data':[]
            };
        {% endfor %}

        let x = [];
        let ids = [];
        let chr = null;
        let min_pos = Infinity;
        let max_pos = -Infinity;
        let _currentChromInfo = _G.chromosomes.idx[_G.chromosomes.nameToIdx[chromosome]];
        let xAxisTickNum = Math.floor(_currentChromInfo.length / 20000000);
        let xAxisTickVals = [];
        let xAxisTickText = [];
        let series = [];
        let effectMinValue = Infinity;
        let effectMaxValue = -Infinity;

        for (let i = 1; i <= xAxisTickNum; i++) {
            xAxisTickVals.push(20000000 * i);
            xAxisTickVals.push((20 * i) + "Mb")
        }


        $.each(effectData.data, function(index, element) {
            //{"A":-0.1224,..."H":1.45,"chr":"4","id":"4_156339889","pos":156.3399}
            let _pos = element.pos * 1000000;
            let m = [];

        {% for s in CONF.PLOT_EFFECT_STRAINS %}
            strainInfo["{{ s['key'] }}"]['data'].push([_pos, element.{{ s['key'] }}]);
            m.push(element.{{ s['key'] }});
        {% endfor %}

            // TODO: might put that at end of loop????
            effectMinValue = Math.min(Math.min.apply(null, m), effectMinValue);
            effectMaxValue = Math.max(Math.max.apply(null, m), effectMaxValue);

            x.push(_pos);
            ids.push(element.id);
            chr = element.chr;
        });

        effectMaxValue = Math.ceil(effectMaxValue);
        effectMinValue = Math.floor(effectMinValue);

        if ((effectMaxValue > 30.0) || (effectMinValue < -30.0)) {
            displayError('Effect Plot Error', 'Effect values appear to be invalid, ' + effectMinValue + ' to ' + effectMaxValue);
            return;
        }

        $.each(strainInfo, function(key, value) {
            let s = {
                type: 'line',
                turboThreshold: 0,
                showInLegend: true,
                data: value['data'],
                name: value['name'],
                color: value['color'],
                lineWidth: 0.5,
                yAxis:'axisEffect',
                stack: 0,
                marker: {
                    symbol: 'circle',
                },
                tooltip: {
                    shared: false,
                    headerFormat:'<b>{series.name}</b>',
                    pointFormatter: function() {
                        return '<br>Effect: ' + this.y + '<br>Position: ' +  (this.x/1000000);
                    }
                }


            };
            series.push(s);
        });

        let _title = '';
        let currentLevel = getCurrentLevel();

        if (currentLevel === 'mrna') {
            _title = _G.current_gene_id + ' (' + _G.current_gene.symbol + ')';
        } else if (currentLevel === 'protein') {
            _title = _G.current_protein_id + ' (' + _G.current_gene.symbol + ')';
        } else {
            _title = _G.current_phenotype_id;
            if (_G.current_phenotype_id !== _G.phenotypes[_G.current_phenotype_id].short_name) {
                _title += (' (' + _G.phenotypes[_G.current_phenotype_id].short_name + ')');
            }
        }

        let lodDataForChromosome = _G.lodDataByChromosome[_currentChromInfo.chromosome];
        let lodData = [];
        let lodMinValue = Infinity;
        let lodMaxValue = -Infinity;

        $.each(lodDataForChromosome, function(index, element) {
            element.x = element.chrPos;
            lodData.push(element);

            lodMinValue = Math.min(element.y, lodMinValue);
            lodMaxValue = Math.max(element.y, lodMaxValue);
        });

        series.push({
            data: lodData,
            // TODO: what should this value be
            turboThreshold: 1000000,
            yAxis:'axisLOD',
            stack: 0,
            name: 'LOD',
            animation: false,
            showInLegend: false,

            tooltip: {
                shared: false,
                headerFormat:'',
                pointFormatter: function() {
                    return '<br>LOD: ' + this.y + '<br>Position: ' +  (this.x/1000000);
                }
            }
        });

        // determine ticks
        // TODO: figure ticks properly
        let effectTickVals = [];
        for (let i = effectMinValue; i <= effectMaxValue; i += 0.5) {
            effectTickVals.push(i);
        }

        let lodTickVals = [];
        let lodMaxTick = Math.ceil(lodMaxValue / 10) * 10;
        for (let i = 0; i <= lodMaxTick; i += 10) {
            lodTickVals.push(i);
        }

        //console.log('lodMaxValue = ', lodMaxValue);
        //console.log('lodMaxTick = ', lodMaxTick);
        //console.log('effectTickVals=', effectTickVals);
        //console.log('lodTickVals=', lodTickVals);

        _G.plots.effect.obj = Highcharts.chart({
            title: {
                text: _title,
            },
            boost: {
                useGPUTranslations: true
            },
            subtitle: {
                text: 'Chromosome ' + chromosome,
                verticalAlign: 'bottom',
            },

            chart: {
                renderTo: _G.plots.effect.el,
                zoomType: 'x'
            },

            xAxis: {
                lineWidth: 0.5,
                gridLineWidth: 1,
                crosshair: true,
            },

            legend: {
                layout: 'vertical',
                align: 'right',
                verticalAlign: 'top',
                floating: false,
                backgroundColor: (Highcharts.theme && Highcharts.theme.legendBackgroundColor) || '#FFFFFF'
            },

            yAxis: [{
                id: 'axisEffect',
                title: {
                    text: 'EFFECT'
                },
                offset: 0,
                height: '70%',
                // TODO: from above, figure out ticks
                //tickPositions: effectTickVals,
                maxPadding: 0,
                labels: {
                    formatter: function() {
                        if (this.isFirst) {
                            return '';
                        }
                        return this.value;
                    }
                },
            }, {
                id: 'axisLOD',
                title: {
                    text: 'LOD'
                },
                offset: 0,
                height: '30%',
                top: '70%',
                maxPadding: 0,
                tickPositions: lodTickVals,
                labels: {
                    formatter: function() {
                        if (this.isLast) {
                            return '';
                        }
                        return this.value;
                    }
                },
                plotLines: [{
                    value: lodMaxTick,
                    color: 'black',
                    zIndex: 5,
                    width: 2
                }]
            }],

            series:series,

            credits: {
                enabled: false // A plugin for highcharts is of course using highcharts
            },

            exporting: {
                buttons: {
                    contextButton: {
                        menuItems: [{
                            text: 'Download JPEG image',
                            onclick: function () {
                                this.exportChartLocal({
                                    filename: _title + '_effect',
                                    type: Highcharts.exporting.MIME_TYPES.JPEG
                                });
                            }
                        }, {
                            text: 'Download PNG image',
                            onclick: function () {
                                this.exportChartLocal({
                                    filename: _title + '_effect',
                                    type: Highcharts.exporting.MIME_TYPES.PNG
                                });
                            },
                            separator: false
                        }, {
                            text: 'Download PDF document',
                            onclick: function () {
                                this.exportChartLocal({
                                    filename: _title + '_effect',
                                    type: Highcharts.exporting.MIME_TYPES.PDF
                                });
                            },
                            separator: false
                        }]
                    }
                }
            }
        });
    }

    /**
     * Plot mediation data
     * @param {Object} mediationData - mediation data
     */
    function plotMediation(mediationData) {
        // probably easier d3 function but just looping for now
        //'LOD': 3.4188, 'chr' : "11", id: "ENSMUSG00000099291", pos:87448734, symbol: "ACEA_U3"

        let n = [];

        $.each(mediationData, function(index, element) {
            if (element.chr in _G.chromosomes.nameToIdx) {
                let chr = _G.chromosomes.idx[_G.chromosomes.nameToIdx[element.chr]];
                let x = chr.start + (element.pos);
                let y = element.LOD;

                let e = element;
                e.x = x;
                e.y = y;
                n.push(e);
            } else {
                //console.log(element);
            }
        });

        let axis_tick_vals = [];
        let axis_tick_text = [];

        // rectangles to display
        let chrom_shapes = [];
        let data_chr = {};

        $.each(_G.chromosomes.idx, function(index, element) {

            axis_tick_vals.push(element.mid);
            axis_tick_text.push(element.name);

            data_chr[element.name] = {};
            data_chr[element.name].data = [];

            let fill_color = '#ffffff';
            if ((index % 2) === 0) {
                fill_color = '#222222';
            }
        });

        //
        // build the vertical chromosome bars
        //
        let chromosomeAxisVals = [];
        let chromosomeAxisText = {};
        let chromosomeBands = [];

        $.each(_G.chromosomes.idx, function(index, element) {

            chromosomeAxisVals.push(element.mid);
            chromosomeAxisText[element.mid] = element.name;

            // TODO: style this
            let fillColor = '#ffffff';
            if ((index % 2) === 1) {
                chromosomeBands.push({
                    color: fillColor,
                    from: element.start,
                    to: element.end
                });
            }
        });

        let _title = '';
        let currentLevel = getCurrentLevel();

        if (currentLevel === 'mrna') {
            _title = _G.current_gene_id + ' (' + _G.current_gene.symbol + ')';
        } else if (currentLevel === 'protein') {
            _title = _G.current_protein_id + ' (' + _G.current_gene.symbol + ')';
        } else {
            _title = _G.current_phenotype_id;
        }

        _G.plots.mediation.obj = Highcharts.chart({
            chart: {
                zoomType: 'x',
                renderTo: _G.plots.mediation.el,
            },

            title: {
                text:_title,
            },

            tooltip: {
                useHTML: true,
                formatter: function () {
                    return '<b>' + this.point.id + '</b>' +
                        '<br/>Symbol: ' + this.point.symbol +
                        '<br/>Chromosome: ' + this.point.chr +
                        '<br/>Location: ' + this.point.pos +
                        '<br/>LOD: ' + this.point.LOD;
                },
                positioner: function(labelWidth, labelHeight, point){
                    let tooltipX, tooltipY;
                    let chart = _G.plots.mediation.obj;
                    if (point.plotX + labelWidth > chart.plotWidth) {
                        tooltipX = point.plotX + chart.plotLeft - labelWidth - 20;
                    } else {
                        tooltipX = point.plotX + chart.plotLeft + 20;
                    }
                    tooltipY = point.plotY + chart.plotTop - 20;
                    return {
                        x: tooltipX,
                        y: tooltipY
                    };
                },
            },
            xAxis: {
                opposite: true,
                tickWidth: 0,
                labels: {
                    formatter: function () {
                        return chromosomeAxisText[this.value];
                    }
                },
                tickPositions: chromosomeAxisVals,
                plotBands: chromosomeBands,
            },

            yAxis: {
                title: {
                    text: 'LOD'
                }
            },

            plotOptions: {
                scatter: {
                    marker: {
                        radius: 5,
                        states: {
                            hover: {
                                enabled: true,
                                lineColor: 'rgb(100,100,100)'
                            }
                        }
                    },
                    states: {
                        hover: {
                            marker: {
                                enabled: false
                            }
                        }
                    },
                }
            },

            series: [{
                type: 'scatter',
                turboThreshold: 0,
                showInLegend: false,
                downsample: {
                    threshold: 0 // 0 disables downsampling
                },
                data: n,
                color: 'rgba(223, 83, 83, .5)',
                marker: {
                      radius:2
                }
            }],

            credits: {
                enabled: false // A plugin for highcharts is of course using highcharts
            },

            exporting: {
                buttons: {
                    contextButton: {
                        menuItems: [{
                            text: 'Download JPEG image',
                            onclick: function () {
                                this.exportChartLocal({
                                    filename: _title + '_mediation',
                                    type: Highcharts.exporting.MIME_TYPES.JPEG
                                });
                            }
                        }, {
                            text: 'Download PNG image',
                            onclick: function () {
                                this.exportChartLocal({
                                    filename: _title + '_mediation',
                                    type: Highcharts.exporting.MIME_TYPES.PNG
                                });
                            },
                            separator: false
                        }, {
                            text: 'Download PDF document',
                            onclick: function () {
                                this.exportChartLocal({
                                    filename: _title + '_mediation',
                                    type: Highcharts.exporting.MIME_TYPES.PDF
                                });
                            },
                            separator: false
                        }]
                    }
                }
            }
        });

        $("#" + _G.plots.mediation.el_wait).html("");
    }

    /**
     * Plot profile data.
     * @param {string} id - the unique identifier
     * @param {Object} profileData - the profile data
     * @param {Array} profiles - array of selected factors
     */
    function plotProfile(id, profileData, profiles) {
        // profiles are covar_factors

        let profile_info = {};

        // create a mapping used for default styling below
        let profile_text_map = {};
        let profile_diet = {};
        let m = [
            DataPointShape.CIRCLE,
            DataPointShape.DIAMOND,
            DataPointShape.SQUARE,
            DataPointShape.CROSS,
            DataPointShape.STAR,
            DataPointShape.X,
            DataPointShape.DOWN_TRIANGLE,
            DataPointShape.UP_TRIANGLE
        ];

        // TODO: hardcoding, needs fixing
        $.each(profileData.data_types, function(key, value) {
            if (key.toLowerCase() === 'sex') {
                profile_text_map['sex'] = key;
            } else if (key.toLowerCase() === 'diet') {
                profile_text_map['diet'] = key;

                if (value.length < 9) {
                    $.each(value, function(i, v) {
                        profile_diet[v] = m[i];
                    });
                }
            }

            profile_info[key] = {levels: value, name: key, kind: 'factor', values: []};
        });

        let samples = [];
        let expression = [];

        $.each(profileData.data, function(index, element) {
            samples.push(element.mouse_id);

            $.each(profileData.data_types, function(key, value) {
                profile_info[key].values.push(element[key]+"");
            });

            expression.push(element.expression);
        });

        let xvar = [];
        $.each(profiles, function(index, element) {
            xvar.push(profile_info[element]);
        });

        let yvar = [{
            kind: 'number',
            name: id,
            values: expression,
        }];

        let plotParams = {
            sampleCount: samples.length,
            xAxis: {
                jitterPx: 25,
                min: null,
                max: null,
                variables: xvar,
                tickLabelRotationDeg: -45
            },
            yAxis: {
                min: null,
                max: null,
                variables: yvar,
                whiskers: {
                    units: 'stderr',
                    dist: 1
                }
            }
        };

        console.log('plotParams=', plotParams);

        _G.fact = {};
        _G.fact.plotParams = plotParams;
        _G.fact.samples = samples;

        // using this so plots have same jitter
        Math.seedrandom('whozapi');

        _G.fact.jsobject = new FactExpPlot({
            svg: d3.select('#fact-svg-chart'),
            pointSizePx: 5,
        });

        // not using mouseOverPoint because it wasn't showing tooltip on initial hover
        //_G.fact.jsobject.mouseOverPoint = function(a, b) {

        /**
         * Here we do some styling.  This is hard coded right now.
         * TODO: FIX hardcoding
         */
        _G.fact.jsobject.postProcessPoint = function(sampleIndex, d3Node) {
            if ('sex' in profile_text_map) {
                d3Node.classed('female', profile_info[profile_text_map['sex']].values[sampleIndex].toLowerCase()[0] === 'f');
                d3Node.classed('male', profile_info[profile_text_map['sex']].values[sampleIndex].toLowerCase()[0] === 'm');
            }

            if ('diet' in profile_text_map) {
                if (profile_diet) {
                    _G.fact.jsobject.setPointShape(d3Node, profile_diet[profile_info[profile_text_map['diet']].values[sampleIndex]]);
                }
            }
        };

        _G.fact.jsobject.renderPlot(plotParams);

        $('#points-group use').each(function () {
            let sample_index = parseInt($(this).attr('sample-index'));
            $(this).popover({
                placement: 'left',
                html: true,
                trigger: 'hover',
                container: 'body',
                content: function () {
                    return _G.fact.samples[sample_index];
                }
            });
        });
    }

    /**
     * Display the gene information.
     * @param {Object} geneData - gene information
     */
    function displayGeneData(geneData) {
        console.log('Displaying gene information:', geneData.id);
        let startFormat = formatMbp(geneData.start/1000000);
        let endFormat = formatMbp(geneData.end/1000000);
        let mbpDesc = 'Mbp';

        if (startFormat < 1) {
            startFormat = geneData.start;
            endFormat = geneData.end;
            mbpDesc = 'bp';
        }

        let synonyms = "";
        if (geneData.synonyms) {
            synonyms = geneData.synonyms.join("<br>");
        }

        let strand = "forward";
        if (geneData.strand === "-") {
            strand = "negative";
        }

        let htmlBody = `                    <div class="row">
                        <div class="col-md-12">
<dl class="dl-horizontal">
                            <dt>ID</dt><dd>${geneData.id||''}
                [
                <a target="_newwin" href="http://www.ensembl.org/id/${geneData.id}">Ensembl <small><i class="fa fa-external-link fa-1" aria-hidden="true"></i></small></a>

                <a target="_newwin" href="http://www.informatics.jax.org/marker/${geneData.id}">MGI <small><i class="fa fa-external-link fa-1" aria-hidden="true"></i></small></a>
                ]
                            </dd>
                            <dt>Symbol</dt><dd>${geneData.symbol||''}</dd>
                            <dt>Chromosome</dt><dd>${geneData.chromosome||''}</dd>
                            <dt>Start Position</dt><dd>${geneData.start||''}</dd>
                            <dt>End Position</dt><dd>${geneData.end||''}</dd>
                            <dt>Strand</dt><dd>${geneData.strand||''}</dd>
                            <dt>Description</dt><dd>${geneData.description||''}</dd>
                            <dt>Synonyms</dt><dd>${synonyms||''}</dd>
                </dl></div></div>`;

        let htmlHeader = `
                    <div class="box-tools pull-left">
                        <button type="button" class="btn btn-box-tool" data-widget="collapse"><i class="fa fa-chevron-right"></i>
                        </button>
                    </div>
                <h3 class="box-title">${geneData.id} <em>${geneData.symbol}</em></h3>
                Chromosome ${geneData.chromosome} (${startFormat} - ${endFormat} ${mbpDesc}, ${strand} strand)
`;

        $("#div_current_gene_information_body").html(htmlBody);
        $("#div_current_gene_information_header").html(htmlHeader);
    }


    /**
     * Perform gene search.
     */
    function findGene() {
        let button = $('#btnGo');
        let term = $('#searchTerm');
        let searchVal = term.val().trim().toUpperCase();

        term.disable(true);
        button.button('loading');

        if (searchVal.length === 0) {
            term.focus();
            term.disable(false);
            button.button('reset');
            return;
        }

        // TODO: should not hard code species
        WHOZ.search(searchVal, {species: 'Mm'}, searchCallback);
    }


    /**
     * Populate the search results.
     */
    function searchCallback() {
        if (WHOZ.response === null) {
            $('#search_results_div').html("");
            return;
        }

        let response = WHOZ.response;

        let tbl = '<table id="search_results_table" class="table table-bordered table-hover table-condensed"></table>';
        $('#search_results_div').html(tbl);

        let searchResults = {};
        for (let d in response.result.matches) {
            let match = response.result.matches[d];
            searchResults[match.ensembl_gene_id] = match;
        }

        let _columns = [];
        let currentLevel = getCurrentLevel();


        if (currentLevel === 'mrna') {
            _columns.push({
                title: 'ID',
                data: 'ensembl_gene_id',
                render: function(data, type, row, meta) {
                    if (type === 'display') {
                        if (_G.genes.gene_ids[data]) {
                            return '<div style="white-space: nowrap"><span id="matchInfo" gene_id="' + data + '" class="glyphicon glyphicon-info-sign"></span> <a href="#" level="' + currentLevel + '" gene_id="' + data + '">' + data + '</a></div>';
                        } else {
                            return '<div style="white-space: nowrap"><span id="matchInfo" gene_id="' + data + '" class="glyphicon glyphicon-info-sign"></span> ' + data + '</div>';
                        }
                    } else {
                        return data;
                    }
                }
            });
        } else {
            _columns.push({
                title: 'ID',
                data: 'ensembl_gene_id',
                render: function(data, type, row, meta) {
                    if (type === 'display') {
                        let ps = [];
                        if (_G.proteins.gene_ids[data]) {
                            let ret = '<div style="white-space: nowrap"><span id="matchInfo" gene_id="' + data + '" class="glyphicon glyphicon-info-sign"></span> ' + data;
                            ret += '</div><div style="padding-left: 20px;">';
                            ret += '<em>';

                            for (p in _G.proteins.gene_ids[data].protein_ids) {
                                ps.push('<a href="#" level="' + currentLevel + '" gene_id="' + data + '" protein_id="' + _G.proteins.gene_ids[data].protein_ids[p] + '">' + _G.proteins.gene_ids[data].protein_ids[p] + '</a>');
                            }

                            return  ret +  ps.join('<br>') + '</em></div>';
                        } else {
                            return '<div style="white-space: nowrap"><span id="matchInfo" gene_id="' + data + '" class="glyphicon glyphicon-info-sign"></span> ' + data +'</div>';
                        }
                    } else {
                        return data;
                    }
                }
            });
        }

        _columns.push({
            title: 'Symbol',
            data: 'symbol'
        });

        _G.searchTable = $('#search_results_table').DataTable({
            data: response.result.matches,
            paging: false,
            order: [],   // prevent automatic ordering
            scrollY: 200,
            searching: false,
            lengthChange: false,
            scrollCollapse: true,
            responsive: true,
            columns: _columns
        });

        if (response.result.num_results === 1) {
            $('#search_results_table_info').html(response.result.matches.length.toLocaleString() + ' result');
        } else if (response.result.matches.num_matches < 100) {
            $('#search_results_table_info').html(response.result.matches.length.toLocaleString() + ' results');
        } else {
            $('#search_results_table_info').html('Showing first ' + response.result.matches.length.toLocaleString() + ' of ' + response.result.num_results.toLocaleString() + ' results');
        }

        $('#search_results_table a').on('click', function (evt) {
            evt.preventDefault();
            let _this = $(this);
            let regressLocal = $("#cond_local").is(":checked");
            selectGeneProtein(_this.attr('gene_id'), _this.attr('protein_id'), _this.attr('level'), regressLocal);
            return false;
        });

        setTimeout(function () {
            $("[id='matchInfo']").each(function () {
                let that = $(this);

                $(this).popover({
                    placement: 'right',
                    html: true,
                    trigger: 'hover',
                    container: 'body',
                    template: '<div class="popover container my-custom-class" role="tooltip"><div class="arrow"></div><h3 class="popover-title"></h3><div class="popover-content"></div></div>',
                    content: function () {
                        let d = searchResults[that[0].getAttribute("gene_id")];
                        let h = '<table class="table table-condensed">';
                        h += '<tr><td><strong>Symbol</strong> </td><td>' + d['symbol'] + '</td></tr>';
                        h += '<tr><td><strong>Location</strong> </td><td>' + d['chromosome'] + ':' + d['position_start'] + '-' + d['position_end'] + '</td></tr>';
                        h += '<tr><td><strong>Name</strong> </td><td>' + d['name'] + '</td></tr>';
                        h += '<tr><td><strong>Synonyms</strong> </td><td>';

                        let s = d['synonyms'];
                        if (s) {
                            if (s.length > 5) {
                                let s2 = s.slice(1, 6);
                                h += s2.join('<br>');
                                h += '<br><i>(' + (s.length - s2.length) + ' more synonyms)</i>';
                            } else {
                                h += s.join('<br>');
                            }
                        }
                        h += '</td></tr>';
                        h += '<tr><td><strong>Match Reason</strong> </td><td>' + d['match_reason'] + '</td></tr>';
                        h += '<tr><td><strong>Match Value</strong> </td><td>' + d['match_value'] + '</td></tr>';
                        h += '</table>';

                        return h;
                    }
                });
            });
        });

        _G.searchResults = searchResults;
        _G.searchResultsLevel = currentLevel;
        $('#btnGo').button('reset');
        $('#searchTerm').disable(false);

        // simulate a click event if the search yields only 1 result
        if ($('#search_results_table a').length === 1) {
            $('#search_results_table a')[0].click();
        }
    }

    /**
     * Handle what should happen upon a new gene selection.
     * @param {string} geneID - the gene identifier
     * @param {string} proteinID - the protein identifier
     * @param {string} level - "mrna" or "protein"
     * @param {boolean} regressLocal - true to regress local or false for default
     */
    function selectGeneProtein(geneID, proteinID, level, regressLocal) {
        blockUser(BLOCK_STATE_SEARCH_SELECTION);
        _G.current_gene_id = geneID;
        _G.current_protein_id = proteinID;

        $('#profile_plot_container_div').removeClass('hidden');
        $('#lod_plot_container_div').removeClass('hidden');
        $('#secondary_plot_container_div').removeClass('hidden');

        let id = geneID;
        if (level === 'protein') {
            id = proteinID;
        }

        d3.queue()
            .defer(downloadGeneData, geneID)
            .defer(downloadLODData, id, level, regressLocal)
            .defer(downloadProfileData, id, level)
            .await(function(error, geneData, lodData, profileData) {
                if (error) {
                    let message = 'Unable to download data for <strong><em>';
                    message += id;
                    message += '</em></strong>.';

                    unblockUser();
                    displayError('API Server Error', message);

                    console.error('error=', error);
                    console.error(message);
                    return;
                }

                _G.current_gene = geneData[0];
                _G.current_gene_id = geneID;
                _G.current_protein_id = proteinID;
                displayGeneData(_G.current_gene);
                resetSecondaryPlots();
                _G.slider.reset();

                plotLOD(lodData.data, id, level);
                plotProfile(id, profileData, getFactorOrder());
                addHistory(_G.current_gene_id, _G.current_protein_id, geneData[0]['symbol'], level);
                buildHistory();
                unblockUser();
            });

    }

    /**
     * Handle what should happen upon a new phenotype selection.
     * @param {string} id - the phenotype identifier
     * @param {string} level - the phenotype level
     * @param {boolean} regressLocal - true to regress local or false for default
     */
    function selectPhenotype(id, level, regressLocal) {
        blockUser(BLOCK_STATE_SEARCH_SELECTION);

        _G.current_phenotype_id = id;

        $('#profile_plot_container_div').removeClass('hidden');
        $('#lod_plot_container_div').removeClass('hidden');
        $('#secondary_plot_container_div').removeClass('hidden');

        d3.queue()
            .defer(downloadPhenotypeData, id)
            .defer(downloadLODData, id, level, regressLocal)
            .defer(downloadProfileData, id, level)
            .await(function(error, phenotypeData, lodData, profileData) {
                if (error) {
                    let message = 'Unable to download data for <strong><em>';
                    message += id;
                    message += '</em></strong>.';

                    unblockUser();
                    displayError('API Server Error', message);

                    console.error('error=', error);
                    console.error(message);

                    return;
                }

                _G.current_phenotype = phenotypeData[0];
                resetSecondaryPlots();
                _G.slider.reset();

                plotLOD(lodData.data, id, level);
                plotProfile(id, profileData, getFactorOrder());
                unblockUser();
            });
    }

    /*******************************************************************************************************************
     **
     ** GENERATE FUNCTIONS
     **
     ******************************************************************************************************************/

    /**
     * Generate the LOD plot by first downloading the data and than plotting it.
     * @param {string} id - unique identifier
     * @param {string} level - the current level
     * @param {boolean} regressLocal - whether to regress locally or not
     */
    function generateLODPlot(id, level, regressLocal) {
        console.log('Generating LOD plot: id=', id, ' level=', level, ' regressLocal=', regressLocal);

        blockUser(BLOCK_STATE_LOD_SCAN);
        resetSecondaryPlots();

        d3.queue()
            .defer(downloadLODData, id, level, regressLocal)
            .await(function(error, lodData) {
                if (error) {
                    let message = 'Unable to download data for <strong><em>';
                    message += id;
                    message += '</em></strong>.';

                    displayError('API Server Error', message);
                    unblockUser();

                    console.error('Error generating LOD plot:', error);
                    console.error(message);

                    return;
                }

                plotLOD(lodData.data, id, level);
                unblockUser();
            });
    }

    /**
     * Generate the SNP association plot by first downloading the data and than plotting it.
     * @param {string} id - unique identifier
     * @param {string} level - current level
     * @param {string} chromosome - chromosome
     * @param {string} position - whether to regress locally or not
     * @param {number} windowSize - size of the window (either side on SNP)
     */
    function generateSNPAssocPlot(id, level, chromosome, position, windowSize) {
        console.log('Generating SNP Assoc plot: id=', id, ' level=', level, ' chromosome=', chromosome, ' position=', position, ' windowSize=', windowSize);

        blockUser(BLOCK_STATE_LOD_SELECTION);

        _G.snpAssocVars = {
            id: id,
            chromosome: chromosome,
            position: position,
        };

        let snpWindow = windowSize / 2;
        let minPos = Math.max(position - snpWindow, 0);
        let maxPos = position + snpWindow;

        d3.queue()
            .defer(downloadSNPAssocData, id, level, chromosome, position, snpWindow)
            .defer(downloadGenesByLocationData, chromosome, minPos, maxPos)
            .await(function(error, snpData, geneData) {
                if (error) {
                    let message = 'Unable to download SNP data for <strong><em>';
                    message += id;
                    message += '</em></strong> on chromosome ';
                    message += '<strong><em>';
                    message += chromosome;
                    message += '</em></strong>';

                    displayError('API Server Error', message);
                    unblockUser();

                    console.error('Error generating SNP assoc plot:', error);
                    console.error(message);

                    return;
                }

                plotSNPAssoc(snpData, geneData);
                unblockUser();
            });
    }

    /**
     * Generate the mediation plot by first downloading the data and than plotting it.
     * @param {string} id - unique identifier
     * @param {string} mid - marker id
     * @param {string} level - current level
     */
    function generateMediationPlot(id, mid, level) {
        console.log('Generating mediation plot: id=', id, ' mid=', mid, ' level=', level);

        blockUser(BLOCK_STATE_LOD_SELECTION);

        d3.queue()
            .defer(downloadMediationData, id, mid, level)
            .await(function(error, data) {
                if (error) {
                    let message = 'Unable to download mediation data for <strong><em>';
                    message += id;
                    message += '</em></strong> on ';
                    message += '<strong><em>';
                    message += mid;
                    message += '</em></strong>';

                    displayError('API Server Error', message);
                    unblockUser();

                    console.error('Error generating mediation plot:', error);
                    console.error(message);

                    return;
                }

                plotMediation(data);
                unblockUser();
            });
    }

    /**
     * Generate the effect plot by first downloading the data and than plotting it.
     * @param {string} id - unique identifier
     * @param {string} level - the current level
     * @param {string} chromosome - the chromosome
     * @param {boolean} blup - whether to do blup analysis or not
     * @param {boolean} regressLocal - true to regress local or false for default
     */
    function generateEffectPlot(id, level, chromosome, blup, regressLocal) {
        console.log('Generating Effect plot: id=', id, ' level=', level, ' chromosome=', chromosome, ' blup=', blup, ' regressLocal=', regressLocal);

        blockUser(BLOCK_STATE_LOD_SELECTION);

        d3.queue()
            .defer(downloadEffectData, id, chromosome, level, blup, regressLocal)
            .await(function(error, data) {
                if (error) {
                    let message = 'Unable to download effect data for <strong><em>';
                    message += id;
                    message += '</em></strong> on chromosome ';
                    message += '<strong><em>';
                    message += chromosome;
                    message += '</em></strong>';

                    displayError('API Server Error', message);
                    unblockUser();

                    console.error('Error generating effect plot:', error);
                    console.error(message);

                    return;
                }

                plotEffect(data, chromosome);
                unblockUser();
            });
    }

    /**
     * Generate the profile plot by first downloading the data and than plotting it.
     * @param {string} id - unique identifier
     * @param {string} level - the current level
     * @param {Array} profiles - array of selected factors
     */
    function generateProfilePlot(id, level, profiles) {
        console.log('Generating Profile plot: id=', id, ' level=', level, ' profiles=', profiles);

        blockUser(BLOCK_STATE_PROFILE_UPDATE);

        d3.queue()
            .defer(downloadProfileData, id, level)
            .await(function(error, data) {
                if (error) {
                    let message = 'Unable to download profile data for <strong><em>';
                    message += id;
                    message += '</em></strong>.';

                    displayError('API Server Error', message);
                    unblockUser();

                    console.error('Error generating profile plot:', error);
                    console.error(message);

                    return;
                }

                plotProfile(id, data, profiles);
                unblockUser();
            });
    }

    /**
     * Initialize some of the controls.
     */
    function init() {
        console.log('initializing...');

        $('#searchTerm').keypress(function(evt) {
            let code = evt.which ? evt.which : evt.keyCode;
            if (code === 13) {
                $(this).blur();
                $('#btnGo').focus().click();
            }
        });

        $('#btnGo').button().on('click', function(evt) {
            evt.preventDefault();
            findGene();
            return false;
        });

        $('#lbl_default').on('click', function(evt) {
            let id = null;
            let currentLevel = getCurrentLevel();

            if (currentLevel === 'mrna') {
                id = _G.current_gene_id;
            } else if (currentLevel === 'protein') {
                id = _G.current_protein_id;
            } else {
                id = _G.current_phenotype_id;
            }

            generateLODPlot(id, currentLevel, false);
        });

        $('#lbl_local').on('click', function(evt) {
            let id = null;
            let currentLevel = getCurrentLevel();

            if (currentLevel === 'mrna') {
                id = _G.current_gene_id;
            } else if (currentLevel === 'protein') {
                id = _G.current_protein_id;
            } else {
                id = _G.current_phenotype_id;
            }

            generateLODPlot(id, currentLevel, true);
        });

        $("#btnFactPlot").on('click', function(evt) {
            let id = null;
            let currentLevel = getCurrentLevel();

            if (currentLevel === 'mrna') {
                id = _G.current_gene_id;
            } else if (currentLevel === 'protein') {
                id = _G.current_protein_id;
            } else {
                id = _G.current_phenotype_id;
            }

            generateProfilePlot(id, currentLevel, getFactorOrder());
        });

        // make sure search tables draw correctly
        $('a[data-toggle="tab"]').on('shown.bs.tab', function (e) {
            if (_G.searchTable) {
                _G.searchTable.draw();
            }
            if (_G.historyTable) {
                _G.historyTable.draw();
            }
        });

        console.log('DONE initializing!')
    }

    /**
     * Initialize the phenotype data.
     * @param {Object} phenotypeData - the phenotype data
     */
    function initPhenotypes(phenotypeData) {
        _G.phenotypes = {};
        let pheno_data = {};

        $.each(phenotypeData.ids, function(index, element) {
            let category = element.category;
            _G.phenotypes[element.data_name] = element;

            if (category in pheno_data) {
                pheno_data[category].push(element);
            } else {
                pheno_data[category] = [element];
            }
        });

        $.typeahead({
            input: '.js-typeahead-pheno',
            maxItem: 100,
            maxItemPerGroup: 100,
            maxLength: 'Infinity',
            minLength: 1,
            display: ["data_name", "short_name"],
            template: "<span>{{'{{'}}short_name{{'}}'}}</span> <small>({{'{{'}}data_name{{'}}'}})</small>",
            dynamic: false,
            order: 'asc',
            hint: true,
            correlativeTemplate: true,
            emptyTemplate: 'No result for "{{'{{'}}query{{'}}'}}"',
            group: {
                key: "category",
                template: "{{'{{'}}category{{'}}'}}"
            },
            source: phenotypeData.ids,
            generateOnLoad: true,
            dropdownFilter: {
                key: "category",
                template: "{{'{{'}}category{{'}}'}}",
                all: 'All'
            },
            callback: {
                onClickAfter: function (node, a, item, event) {
                    event.preventDefault();
                    // href key gets added inside item from options.href configuration
                    console.log(item);

                    selectPhenotype(item.data_name, getCurrentLevel(), $("#cond_local").is(":checked"));
                },
                onSubmit: function(node, form, item, event) {
                    event.preventDefault();
                }
            },
            debug: true
        });
    }

    /**
     * Get the current level.
     * @return {string} the current level
     */
    function getCurrentLevel() {
        let val = $("[id^='lbl_level'] :checked").val();

        if (val === undefined) {
            return _G.data_levels[0];
        }

        return val.toLowerCase();
    }

    /**
     * Initialize the elements on the page.
     */
    $().ready(function () {
        init();

        WHOZ = new whoz({
            search_limit: 100,
        });

        console.log('Starting Queue...');

        d3.queue()
            .defer(downloadOptions)
            .defer(downloadChromosomes)
            .defer(downloadGenes)
            .defer(downloadProteins)
            .defer(downloadPhenotypes)
            .await(function(error, options, chromosomes, genes, proteins, phenotypes) {
                if (error) {
                    console.error('INITIALIZATION ERROR: ', error);
                    return;
                }

                _G.isPhenoExperiment = false;

                if (phenotypes.ids.length) {
                    _G.isPhenoExperiment = true;
                    $('#search_pheno').removeClass('hidden');
                    $('#mediation_div').addClass('hidden');
                } else {
                    $('#search_gene').removeClass('hidden');
                }

                initPhenotypes(phenotypes);

                _G.covar_factors = options.covar_factors;
                _G.max_lod_points = options.num_snps;
                _G.data_levels = options.data_levels;
                if (!(_G.data_levels instanceof Array)) {
                    // will be a string
                    _G.data_levels = [options.data_levels];
                }


                _G.chromosomes = {'idx': chromosomes.chromosomes, 'nameToIdx': {}};
                $.each(chromosomes.chromosomes, function(index, element) {
                    _G.chromosomes['nameToIdx'][element.name] = index;
                });

                _G.chromosomes['totalLength'] = _G.chromosomes.idx[_G.chromosomes.idx.length-1].end;

                let simple_dict = {'mrna': 'mRNA',
                                   'protein': 'Protein'};

                // data levels could be an array of levels or a string of just 1
                if ((_G.data_levels instanceof Array) && (_G.data_levels.length > 1)) {
                    $.each(_G.data_levels, function(index, element) {
                        let active = '';
                        let checked = '';

                        if (index === 0) {
                            active = 'active';
                            checked = 'checked';
                        }

                        if (_G.isPhenoExperiment) {
                            $('#div_btn_levels').append('<label id="lbl_level_' + index + '" class="btn btn-sm btn-flat btn-primary ' + active + '"><input type="radio" name="radio_level" id="radio_level" value="' + element + '" autocomplete="off" ' + checked + '> ' + element + '</label>');
                            $('#lbl_level_' + index).on("click", function() {
                                let regressLocal = $("#cond_local").is(":checked");
                                if (window.Typeahead['.js-typeahead-pheno'].item.data_name !== undefined) {
                                    selectPhenotype(window.Typeahead['.js-typeahead-pheno'].item.data_name, element, regressLocal);
                                }
                            });
                        } else {
                            $('#div_btn_levels').append('<label id="lbl_level_' + index + '" class="btn btn-sm btn-flat btn-primary ' + active + '"><input type="radio" name="radio_level" id="radio_level" value="' + simple_dict[element] + '" autocomplete="off" ' + checked + '> ' + simple_dict[element] + '</label>');
                            $('#lbl_level_' + index).on("click", function() {
                                findGene();
                            });
                        }
                    });
                }

                let _gene_ids = {};


                $.each(genes.ids, function(index, element) {
                    if (!(element.gene_id in _gene_ids)) {
                        _gene_ids[element.gene_id] = {'id': element.gene_id}
                    }
                });

                _G.genes = {gene_ids:_gene_ids};
                _gene_ids = {};

                let _protein_ids = [];
                $.each(proteins.ids, function(index, element) {
                    if (!(element.gene_id in _gene_ids)) {
                        _gene_ids[element.gene_id] = {'id': element.gene_id, 'protein_ids':[]}
                    }
                    _protein_ids.push(element.protein_id);
                    _gene_ids[element.gene_id]['protein_ids'].push(element.protein_id);
                });

                _G.proteins = {gene_ids:_gene_ids, proteins_ids: _protein_ids};

                buildHistory();

                {% if search_term %}
                console.log('Search term passed in, calling findGene()...');
                findGene();
                {% else %}
                //console.log('Search term NOT passed in');
                {% endif %}

                _G.orderCount = _G.covar_factors.length;

                $('#factor-view-select').multiselect({
                    onChange: function(option, checked) {
                        if (checked) {
                            _G.orderCount++;
                            $(option).attr('order', _G.orderCount);
                        }
                        else {
                            $(option).attr('order', '');
                        }
                    },
                    maxHeight: 400,
                    dropUp: false,
                    buttonWidth: '100%',
                    buttonText: function(options) {
                        if (options.length === 0) {
                            return 'None selected';
                        }
                        else {
                            let selected = [];

                            options.each(function() {
                                selected.push([$(this).text(), $(this).attr('order')]);
                            });

                            selected.sort(function(a, b) {
                                return a[1] - b[1];
                            });

                            let text = '';
                            for (let i = 0; i < selected.length; i++) {
                                text += ((i+1) + ": " + selected[i][0] + ', ');
                            }

                            return text.substr(0, text.length -2);
                        }
                    },
                });

                let selectOptions = [];

                $.each(_G.covar_factors, function(index, element) {
                    selectOptions.push({
                        label: element.display_name,
                        title: element.display_name,
                        value: element.column_name,
                        order: index,
                        selected: true,
                    });

                });

                $('#factor-view-select').multiselect('dataprovider', selectOptions);

                $('#range_1').ionRangeSlider({
                      min: 1000,
                      max: _G.max_lod_points,
                      from: 10000,
                      type: 'single',
                      step: 100,
                      prettify: function (num) {
                          return num + "";
                      },
                      onChange: function(data) {
                          _G.plots.lod.obj.series[0].options.downsample.threshold = data.from;
                          _G.plots.lod.obj.series[0].setData(_G.newLodData);
                      }
                });

                _G.slider = $('#range_1').data('ionRangeSlider');

                $('#rangeSNPWindow').ionRangeSlider({
                    type: 'single',
                    values: [1000000, 2000000, 3000000, 4000000, 5000000],
                    prettify: function (num) {
                        return num.toLocaleString() + "";
                    },
                });

                $('#btnSNPAssocRefresh').on('click', function() {
                    generateSNPAssocPlot(_G.snpAssocVars.id, getCurrentLevel(), _G.snpAssocVars.chromosome, _G.snpAssocVars.position, _G.sliderSNP.result.from_value);
                });

                _G.sliderSNP = $('#rangeSNPWindow').data('ionRangeSlider');
                _G.sliderSNP.reset();
        });
    });
    //-->
</script>

</body>
</html>


